{
  "metadata": {
    "kernelspec": {
      "language": "python",
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.10.12",
      "mimetype": "text/x-python",
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "pygments_lexer": "ipython3",
      "nbconvert_exporter": "python",
      "file_extension": ".py"
    },
    "kaggle": {
      "accelerator": "none",
      "dataSources": [
        {
          "sourceId": 230200333,
          "sourceType": "kernelVersion"
        }
      ],
      "dockerImageVersionId": 30918,
      "isInternetEnabled": true,
      "language": "python",
      "sourceType": "notebook",
      "isGpuEnabled": false
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat_minor": 0,
  "nbformat": 4,
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "## Pipeline Corrosion_protein_classification\n",
        "on Sequence Analysis and Functional Prediction"
      ],
      "metadata": {
        "id": "TSaiUYII_FkF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "os.path.exists('/content/drive/MyDrive')"
      ],
      "metadata": {
        "_uuid": "8f2839f25d086af736a60e9eeb907d3b93b6e0e5",
        "_cell_guid": "b1076dfc-b9ad-4769-8c92-a6c4dae69d19",
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-03-28T13:27:08.859814Z",
          "iopub.execute_input": "2025-03-28T13:27:08.860102Z",
          "iopub.status.idle": "2025-03-28T13:27:08.870718Z",
          "shell.execute_reply.started": "2025-03-28T13:27:08.860075Z",
          "shell.execute_reply": "2025-03-28T13:27:08.869390Z"
        },
        "id": "Lemg3Ilw_FkG",
        "outputId": "b56cf71d-a369-49bb-baa2-1065ea65fa40"
      },
      "outputs": [
        {
          "execution_count": 1,
          "output_type": "execute_result",
          "data": {
            "text/plain": "False"
          },
          "metadata": {}
        }
      ],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# Colab specific\n",
        "from google.colab import drive\n",
        "from google.colab import files\n",
        "import os\n",
        "\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "#change the path\n",
        "os.chdir('/content/drive/MyDrive/MIC/data_picrust')"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-03-28T13:27:08.872742Z",
          "iopub.execute_input": "2025-03-28T13:27:08.873293Z",
          "iopub.status.idle": "2025-03-28T13:27:08.996056Z",
          "shell.execute_reply.started": "2025-03-28T13:27:08.873251Z",
          "shell.execute_reply": "2025-03-28T13:27:08.994428Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DwAqBypS_FkG",
        "outputId": "df1de336-8651-4a43-f9c8-bf7665d1c949"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "execution_count": 1
    },
    {
      "cell_type": "code",
      "source": [
        "# Making sure to use same python version for compatibility\n",
        "!sudo apt-get update -y\n",
        "!sudo apt-get install python3.10\n",
        "!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.10 1\n",
        "!python --version"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-03-28T13:31:57.946204Z",
          "iopub.execute_input": "2025-03-28T13:31:57.946532Z",
          "iopub.status.idle": "2025-03-28T13:32:30.395315Z",
          "shell.execute_reply.started": "2025-03-28T13:31:57.946504Z",
          "shell.execute_reply": "2025-03-28T13:32:30.394064Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MpGDnDv5_FkG",
        "outputId": "72aed3fb-0e5c-4793-ed1e-f60081f5782a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\n",
            "Get:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:8 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Get:9 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,381 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:12 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,691 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,045 kB]\n",
            "Get:14 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,796 kB]\n",
            "Get:15 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [3,892 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [55.7 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,538 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [4,049 kB]\n",
            "Get:19 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,737 kB]\n",
            "Get:20 http://security.ubuntu.com/ubuntu jammy-security/multiverse amd64 Packages [47.7 kB]\n",
            "Get:21 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,239 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [35.2 kB]\n",
            "Fetched 29.9 MB in 12s (2,448 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "python3.10 is already the newest version (3.10.12-1~22.04.9).\n",
            "python3.10 set to manually installed.\n",
            "0 upgraded, 0 newly installed, 0 to remove and 40 not upgraded.\n",
            "Python 3.11.11\n"
          ]
        }
      ],
      "execution_count": 2
    },
    {
      "cell_type": "code",
      "source": [
        "# Set up memory footprint support libraries\n",
        "!ln -sf /opt/bin/nvidia-smi /usr/bin/nvidia-smi\n",
        "!pip install gputil\n",
        "!pip install psutil\n",
        "%pip install humanize\n",
        "%pip install memory_profiler\n",
        "import psutil\n",
        "import humanize\n",
        "import os\n",
        "import GPUtil as GPU\n",
        "GPUs = GPU.getGPUs()\n",
        "from psutil import virtual_memory\n",
        "ram_gb = virtual_memory().total / 1e9\n",
        "print('Runtime has {:.1f} gigabytes of available RAM\\n'.format(ram_gb))\n",
        "\n",
        "if ram_gb < 20:\n",
        "  print('Not using a high-RAM runtime')\n",
        "else:\n",
        "  print('Using a high-RAM runtime!')"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-03-28T13:32:36.128513Z",
          "iopub.execute_input": "2025-03-28T13:32:36.128904Z",
          "iopub.status.idle": "2025-03-28T13:32:58.007809Z",
          "shell.execute_reply.started": "2025-03-28T13:32:36.128873Z",
          "shell.execute_reply": "2025-03-28T13:32:58.006473Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Lrdi1lHT_FkG",
        "outputId": "66c5f8f5-22f0-4141-f67c-77133b362cfc"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting gputil\n",
            "  Downloading GPUtil-1.4.0.tar.gz (5.5 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Building wheels for collected packages: gputil\n",
            "  Building wheel for gputil (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for gputil: filename=GPUtil-1.4.0-py3-none-any.whl size=7392 sha256=7532f5d0cb90378d9738f1902c2521a90cbf0992b9e8fb732acc86722a63e7df\n",
            "  Stored in directory: /root/.cache/pip/wheels/2b/4d/8f/55fb4f7b9b591891e8d3f72977c4ec6c7763b39c19f0861595\n",
            "Successfully built gputil\n",
            "Installing collected packages: gputil\n",
            "Successfully installed gputil-1.4.0\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (5.9.5)\n",
            "Requirement already satisfied: humanize in /usr/local/lib/python3.11/dist-packages (4.12.1)\n",
            "Collecting memory_profiler\n",
            "  Downloading memory_profiler-0.61.0-py3-none-any.whl.metadata (20 kB)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from memory_profiler) (5.9.5)\n",
            "Downloading memory_profiler-0.61.0-py3-none-any.whl (31 kB)\n",
            "Installing collected packages: memory_profiler\n",
            "Successfully installed memory_profiler-0.61.0\n",
            "Runtime has 13.6 gigabytes of available RAM\n",
            "\n",
            "Not using a high-RAM runtime\n"
          ]
        }
      ],
      "execution_count": 3
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install psutil\n",
        "import psutil\n",
        "!pip install biopython\n",
        "from IPython import get_ipython\n",
        "from IPython.display import display\n",
        "!pip install biom-format\n",
        "%pip install umap-learn\n",
        "!pip install lxml pandas\n",
        "!pip install pyarrow\n",
        "!pip install kneed\n",
        "!pip install scipy\n",
        "%pip install \"dask[complete]\"\n",
        "!pip install fuzzywuzzy\n",
        "!pip install python-Levenshtein"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-03-28T13:32:58.009116Z",
          "iopub.execute_input": "2025-03-28T13:32:58.009427Z",
          "iopub.status.idle": "2025-03-28T13:34:19.313693Z",
          "shell.execute_reply.started": "2025-03-28T13:32:58.009397Z",
          "shell.execute_reply": "2025-03-28T13:34:19.312140Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "02-vZuFo_FkG",
        "outputId": "2d41f04e-c784-4943-a910-179d2b0f97c3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (5.9.5)\n",
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m21.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.85\n",
            "Collecting biom-format\n",
            "  Downloading biom-format-2.1.16.tar.gz (11.7 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.7/11.7 MB\u001b[0m \u001b[31m36.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.11/dist-packages (from biom-format) (8.1.8)\n",
            "Requirement already satisfied: numpy>=1.9.2 in /usr/local/lib/python3.11/dist-packages (from biom-format) (2.0.2)\n",
            "Requirement already satisfied: scipy>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from biom-format) (1.14.1)\n",
            "Requirement already satisfied: pandas>=0.20.0 in /usr/local/lib/python3.11/dist-packages (from biom-format) (2.2.2)\n",
            "Requirement already satisfied: h5py in /usr/local/lib/python3.11/dist-packages (from biom-format) (3.13.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas>=0.20.0->biom-format) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas>=0.20.0->biom-format) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas>=0.20.0->biom-format) (2025.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas>=0.20.0->biom-format) (1.17.0)\n",
            "Building wheels for collected packages: biom-format\n",
            "  Building wheel for biom-format (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for biom-format: filename=biom_format-2.1.16-cp311-cp311-linux_x86_64.whl size=12182991 sha256=4c956d0e411c4d6d852370543e52792220f384ddb97160555ed2ae544db01d3d\n",
            "  Stored in directory: /root/.cache/pip/wheels/a5/6b/58/a879e8fbae2479a3d1a68719f3a062fe62701d6494f1b74f5e\n",
            "Successfully built biom-format\n",
            "Installing collected packages: biom-format\n",
            "Successfully installed biom-format-2.1.16\n",
            "Requirement already satisfied: umap-learn in /usr/local/lib/python3.11/dist-packages (0.5.7)\n",
            "Requirement already satisfied: numpy>=1.17 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (2.0.2)\n",
            "Requirement already satisfied: scipy>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (1.14.1)\n",
            "Requirement already satisfied: scikit-learn>=0.22 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (1.6.1)\n",
            "Requirement already satisfied: numba>=0.51.2 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (0.60.0)\n",
            "Requirement already satisfied: pynndescent>=0.5 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (0.5.13)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from umap-learn) (4.67.1)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.11/dist-packages (from numba>=0.51.2->umap-learn) (0.43.0)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.11/dist-packages (from pynndescent>=0.5->umap-learn) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn>=0.22->umap-learn) (3.6.0)\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.11/dist-packages (5.3.1)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (2.2.2)\n",
            "Requirement already satisfied: numpy>=1.23.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas) (2025.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Requirement already satisfied: pyarrow in /usr/local/lib/python3.11/dist-packages (18.1.0)\n",
            "Collecting kneed\n",
            "  Downloading kneed-0.8.5-py3-none-any.whl.metadata (5.5 kB)\n",
            "Requirement already satisfied: numpy>=1.14.2 in /usr/local/lib/python3.11/dist-packages (from kneed) (2.0.2)\n",
            "Requirement already satisfied: scipy>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from kneed) (1.14.1)\n",
            "Downloading kneed-0.8.5-py3-none-any.whl (10 kB)\n",
            "Installing collected packages: kneed\n",
            "Successfully installed kneed-0.8.5\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (1.14.1)\n",
            "Requirement already satisfied: numpy<2.3,>=1.23.5 in /usr/local/lib/python3.11/dist-packages (from scipy) (2.0.2)\n",
            "Requirement already satisfied: dask[complete] in /usr/local/lib/python3.11/dist-packages (2024.12.1)\n",
            "Requirement already satisfied: click>=8.1 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (8.1.8)\n",
            "Requirement already satisfied: cloudpickle>=3.0.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (3.1.1)\n",
            "Requirement already satisfied: fsspec>=2021.09.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (2025.3.0)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (24.2)\n",
            "Requirement already satisfied: partd>=1.4.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (1.4.2)\n",
            "Requirement already satisfied: pyyaml>=5.3.1 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (6.0.2)\n",
            "Requirement already satisfied: toolz>=0.10.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (0.12.1)\n",
            "Requirement already satisfied: importlib_metadata>=4.13.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (8.6.1)\n",
            "Requirement already satisfied: pyarrow>=14.0.1 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (18.1.0)\n",
            "Collecting lz4>=4.3.2 (from dask[complete])\n",
            "  Downloading lz4-4.4.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.8 kB)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.11/dist-packages (from importlib_metadata>=4.13.0->dask[complete]) (3.21.0)\n",
            "Requirement already satisfied: locket in /usr/local/lib/python3.11/dist-packages (from partd>=1.4.0->dask[complete]) (1.0.0)\n",
            "Requirement already satisfied: numpy>=1.24 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (2.0.2)\n",
            "Requirement already satisfied: pandas>=2.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (2.2.2)\n",
            "Requirement already satisfied: dask-expr<1.2,>=1.1 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (1.1.21)\n",
            "Requirement already satisfied: distributed==2024.12.1 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (2024.12.1)\n",
            "Requirement already satisfied: bokeh>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (3.6.3)\n",
            "Requirement already satisfied: jinja2>=2.10.3 in /usr/local/lib/python3.11/dist-packages (from dask[complete]) (3.1.6)\n",
            "Requirement already satisfied: msgpack>=1.0.2 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (1.1.0)\n",
            "Requirement already satisfied: psutil>=5.8.0 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (5.9.5)\n",
            "Requirement already satisfied: sortedcontainers>=2.0.5 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (2.4.0)\n",
            "Requirement already satisfied: tblib>=1.6.0 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (3.0.0)\n",
            "Requirement already satisfied: tornado>=6.2.0 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (6.4.2)\n",
            "Requirement already satisfied: urllib3>=1.26.5 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (2.3.0)\n",
            "Requirement already satisfied: zict>=3.0.0 in /usr/local/lib/python3.11/dist-packages (from distributed==2024.12.1->dask[complete]) (3.0.0)\n",
            "Requirement already satisfied: contourpy>=1.2 in /usr/local/lib/python3.11/dist-packages (from bokeh>=3.1.0->dask[complete]) (1.3.1)\n",
            "Requirement already satisfied: pillow>=7.1.0 in /usr/local/lib/python3.11/dist-packages (from bokeh>=3.1.0->dask[complete]) (11.1.0)\n",
            "Requirement already satisfied: xyzservices>=2021.09.1 in /usr/local/lib/python3.11/dist-packages (from bokeh>=3.1.0->dask[complete]) (2025.1.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2>=2.10.3->dask[complete]) (3.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas>=2.0->dask[complete]) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas>=2.0->dask[complete]) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas>=2.0->dask[complete]) (2025.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas>=2.0->dask[complete]) (1.17.0)\n",
            "Downloading lz4-4.4.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m12.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: lz4\n",
            "Successfully installed lz4-4.4.3\n",
            "Collecting fuzzywuzzy\n",
            "  Downloading fuzzywuzzy-0.18.0-py2.py3-none-any.whl.metadata (4.9 kB)\n",
            "Downloading fuzzywuzzy-0.18.0-py2.py3-none-any.whl (18 kB)\n",
            "Installing collected packages: fuzzywuzzy\n",
            "Successfully installed fuzzywuzzy-0.18.0\n",
            "Collecting python-Levenshtein\n",
            "  Downloading python_levenshtein-0.27.1-py3-none-any.whl.metadata (3.7 kB)\n",
            "Collecting Levenshtein==0.27.1 (from python-Levenshtein)\n",
            "  Downloading levenshtein-0.27.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.6 kB)\n",
            "Collecting rapidfuzz<4.0.0,>=3.9.0 (from Levenshtein==0.27.1->python-Levenshtein)\n",
            "  Downloading rapidfuzz-3.12.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
            "Downloading python_levenshtein-0.27.1-py3-none-any.whl (9.4 kB)\n",
            "Downloading levenshtein-0.27.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (161 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m161.7/161.7 kB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading rapidfuzz-3.12.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m45.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: rapidfuzz, Levenshtein, python-Levenshtein\n",
            "Successfully installed Levenshtein-0.27.1 python-Levenshtein-0.27.1 rapidfuzz-3.12.2\n"
          ]
        }
      ],
      "execution_count": 4
    },
    {
      "cell_type": "code",
      "source": [
        "# Standard library imports\n",
        "import os\n",
        "import sys\n",
        "import ast\n",
        "import subprocess\n",
        "import logging\n",
        "import time\n",
        "from datetime import datetime\n",
        "import shutil\n",
        "from io import StringIO\n",
        "from pathlib import Path\n",
        "import re\n",
        "import json\n",
        "# Data processing and analysis\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import openpyxl\n",
        "import seaborn as sns\n",
        "import networkx as nx\n",
        "import plotly.express as px\n",
        "import plotly.graph_objects as go\n",
        "from plotly.subplots import make_subplots\n",
        "import plotly.io as pio\n",
        "import matplotlib\n",
        "matplotlib.use('Agg')\n",
        "import matplotlib.pyplot as plt\n",
        "%matplotlib inline\n",
        "from matplotlib.colors import to_rgba, LinearSegmentedColormap\n",
        "import matplotlib.patches as mpatches\n",
        "# Machine learning and statistical analysis\n",
        "from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler\n",
        "from sklearn.decomposition import PCA, NMF\n",
        "from sklearn.cluster import AgglomerativeClustering, KMeans\n",
        "from sklearn.manifold import TSNE\n",
        "import umap\n",
        "import scipy\n",
        "from scipy import stats\n",
        "from scipy.cluster import hierarchy\n",
        "import scipy.cluster.hierarchy as sch\n",
        "from statsmodels.stats.multitest import multipletests\n",
        "from scipy.spatial.distance import pdist\n",
        "from scipy.stats import spearmanr, kruskal, mannwhitneyu\n",
        "from kneed import KneeLocator\n",
        "from scipy.signal import savgol_filter\n",
        "from joblib import Parallel, delayed\n",
        "\n",
        "# Bioinformatics\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from biom import Table, load_table\n",
        "from biom.util import biom_open\n",
        "\n",
        "# Web and data retrieval\n",
        "import requests\n",
        "import xml.etree.ElementTree as ET\n",
        "from lxml import etree\n",
        "\n",
        "# Utility libraries\n",
        "import gzip\n",
        "import random\n",
        "from natsort import natsorted\n",
        "from typing import Dict, List, Tuple, Set, Optional\n",
        "import pickle\n",
        "import gc\n",
        "import joblib\n",
        "import h5py\n",
        "import os\n",
        "os.environ['DISPLAY'] = ':0'"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-03-28T13:34:19.316120Z",
          "iopub.execute_input": "2025-03-28T13:34:19.316559Z"
        },
        "id": "tGsaOWH8_FkH"
      },
      "outputs": [],
      "execution_count": 5
    },
    {
      "cell_type": "code",
      "source": [
        "# Set up logging\n",
        "logging.basicConfig(\n",
        "    level=logging.INFO,\n",
        "    format='%(asctime)s - %(levelname)s - %(message)s'\n",
        ")\n",
        "\n",
        "# Directory Structure Definitions\n",
        "SIMPLE_BASE = {\n",
        "    'known': 'simple_known_mic',\n",
        "    'other': 'simple_candidate_mic'\n",
        "}\n",
        "\n",
        "DETAILED_BASE = {\n",
        "    'known': 'detailed_known_mic',\n",
        "    'pure_checked': 'detailed_pure_checked_mic',\n",
        "    'pure_core': 'detailed_pure_core_mic',\n",
        "    'checked_core': 'detailed_checked_core_mic'\n",
        "}\n",
        "\n",
        "SUBDIRS = [\n",
        "    'EC_predictions',\n",
        "    'pathway_predictions',\n",
        "    'KO_predictions',\n",
        "    'other_picrust_files'\n",
        "]\n",
        "\n",
        "\n",
        "# Base Paths\n",
        "if \"google.colab\" in sys.modules:\n",
        "    base_dir = Path(\"/content/drive/MyDrive/MIC/data_picrust\")\n",
        "else:\n",
        "    base_dir = Path(\"/home/beatriz/MIC/2_Micro/data_picrust\")\n",
        "\n",
        "#base dir for small files to git\n",
        "base_dir = Path(\"/home/beatriz/MIC/2_Micro/data_picrust\")\n",
        "\n",
        "abundance_excel= Path(\"/home/beatriz/MIC/2_Micro/data_Ref/merged_to_sequence.xlsx\")\n",
        "fasta_file_final = Path(\"/home/beatriz/MIC/2_Micro/data_qiime/results_match_gg/final_sequences_gg.fasta\")\n",
        "aligned_fasta = Path(\"/home/beatriz/MIC/2_Micro/data_qiime/results_match_gg/aligned-dna-sequences_gg.fasta\")\n",
        "\n",
        "# Create output directory if it doesn't exist\n",
        "output_base = base_dir / \"output_base\"\n",
        "output_base.mkdir(parents=True, exist_ok=True)\n",
        "# large galaxies input and output #large size dir for large files hosted instead in kaggle\n",
        "large_dir = Path(\"/home/beatriz/MIC\")\n",
        "large_dir.mkdir(parents=True, exist_ok=True)\n",
        "# databases\n",
        "db_dir = large_dir / \"Databases\"\n",
        "# input galaxies and uniprots\n",
        "input_galaxy = large_dir / \"data_galaxies\"\n",
        "# Directory to output large files # eccontris, compilated db\n",
        "output_large = large_dir / \"output_large\"\n",
        "output_large.mkdir(parents=True, exist_ok=True)\n",
        "\n",
        "# for colab\n",
        "# Create output directory if it doesn't exist\n",
        "base_dir = Path(\"/content/drive/MyDrive/MIC/data_picrust/\")\n",
        "base_dir.mkdir(parents=True, exist_ok=True)\n",
        "abundance_excel= Path(\"/content/drive/MyDrive/MIC/data_picrust/merged_to_sequence.xlsx\")\n",
        "fasta_file_final = Path(\"/content/drive/MyDrive/MIC/data_picrust/final_sequences_gg.fasta\")\n",
        "aligned_fasta = Path(\"/content/drive/MyDrive/MIC/data_picrust/aligned-dna-sequences_gg.fasta\")\n",
        "\n",
        "output_base = base_dir  # Separate output directory\n",
        "output_base.mkdir(parents=True, exist_ok=True)\n",
        "large_dir = Path(\"/content/drive/MyDrive/MIC/\")\n",
        "large_dir.mkdir(parents=True, exist_ok=True)\n",
        "db_dir = Path(\"/content/drive/MyDrive/MIC/Databases\")\n",
        "db_dir.mkdir(parents=True, exist_ok=True)\n",
        "input_galaxy = large_dir / \"data_galaxies\"\n",
        "# Directory to output large files # eccontris, compilated db\n",
        "output_large = large_dir / \"output_large\"\n",
        "output_large.mkdir(parents=True, exist_ok=True)\n",
        "'''\n",
        "# For Kaggle work\n",
        "# Input datasets (read-only in Kaggle)\n",
        "base_dir = Path(\"/kaggle/input/new-picrust\") #base dir for small files to git /kaggle/input/new-picrust\n",
        "\n",
        "# Files in small input directory\n",
        "abundance_excel= base_dir / \"merged_to_sequence.xlsx\" # inside input small sizes input\n",
        "fasta_file_final = base_dir  / \"final_sequences_gg.fasta\" # inside input small sizes\n",
        "\n",
        "# Output for small files has to be changed for vscode no to push it to git\n",
        "output_base = Path(\"/kaggle/working/output_base\")\n",
        "output_base.mkdir(parents=True, exist_ok=True)\n",
        "#datasets large galaxies and databases\n",
        "db_dir = Path(\"/kaggle/input/databases/Databases\")\n",
        "input_galaxy = Path(\"/kaggle/input/data-galaxies\")\n",
        "\n",
        "# Directory to output large files # eccontris, compilated db\n",
        "large_dir =  Path(\"/kaggle/working/\")\n",
        "\n",
        "# Directory to output large files # eccontris, compilated db\n",
        "output_large = large_dir / \"output_large\"\n",
        "output_large.mkdir(parents=True, exist_ok=True)'''"
      ],
      "metadata": {
        "trusted": true,
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 226
        },
        "id": "ge1xuO4H_FkH",
        "outputId": "fc7c38b1-4fb9-4f65-b73b-e7f688a813c6"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\n# For Kaggle work\\n# Input datasets (read-only in Kaggle)\\nbase_dir = Path(\"/kaggle/input/new-picrust\") #base dir for small files to git /kaggle/input/new-picrust\\n\\n# Files in small input directory\\nabundance_excel= base_dir / \"merged_to_sequence.xlsx\" # inside input small sizes input\\nfasta_file_final = base_dir  / \"final_sequences_gg.fasta\" # inside input small sizes\\n\\n# Output for small files has to be changed for vscode no to push it to git\\noutput_base = Path(\"/kaggle/working/output_base\")\\noutput_base.mkdir(parents=True, exist_ok=True)\\n#datasets large galaxies and databases\\ndb_dir = Path(\"/kaggle/input/databases/Databases\")\\ninput_galaxy = Path(\"/kaggle/input/data-galaxies\")\\n\\n# Directory to output large files # eccontris, compilated db\\nlarge_dir =  Path(\"/kaggle/working/\")\\n\\n# Directory to output large files # eccontris, compilated db\\noutput_large = large_dir / \"output_large\"\\noutput_large.mkdir(parents=True, exist_ok=True)'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "execution_count": 6
    },
    {
      "cell_type": "code",
      "source": [
        "# Integrated taxa from origin genus as headers with levels 6 for the genera, 7 for the GID, muss be cleaned\n",
        "Integrated_T = pd.read_excel(abundance_excel, sheet_name='core_check_usual_taxa', header=[0,1,2,3,4,5,6,7], engine ='openpyxl')\n",
        "# Drop first row (index 0) and first column in one chain\n",
        "Integrated_T = Integrated_T.drop(index=0).drop(Integrated_T.columns[0], axis=1)\n",
        "Integrated_T= Integrated_T.astype({'Sites': str})\n",
        "Integrated_T['Sites'] = Integrated_T['Sites'].fillna('Source')\n",
        "# Remove 'Unnamed' level names\n",
        "Integrated_T.columns = Integrated_T.columns.map(lambda x: tuple('' if 'Unnamed' in str(level) else level for level in x))\n",
        "# Changing dtypes to category whiles respecting structure\n",
        "Integrated_T[\"Category\"] = Integrated_T[\"Category\"].astype(\"Int64\")\n",
        "Integrated_T= Integrated_T.set_index(\"Sites\")\n",
        "pre_Integrated = Integrated_T.T"
      ],
      "metadata": {
        "trusted": true,
        "id": "ZsMx3Uvr_FkH"
      },
      "outputs": [],
      "execution_count": 7
    },
    {
      "cell_type": "code",
      "source": [
        "# Load colab already has corrected the name and put identifiers at this point and category included\n",
        "ECcontri_Uniprot_enriched = pd.read_parquet(output_large / 'ECcontri_Uniprot_enriched.parquet')"
      ],
      "metadata": {
        "trusted": true,
        "id": "hwZVj7ta_FkH"
      },
      "outputs": [],
      "execution_count": 8
    },
    {
      "cell_type": "code",
      "source": [
        "# Define category dict outside\n",
        "category_dict = Integrated_T.T.iloc[0, 0:-1].to_dict()\n",
        "\n",
        "# Define colors and categories\n",
        "category_colors = {1: '#008800',  # Dark green\n",
        "                   2: '#FF8C00',  # Dark orange\n",
        "                   3: '#FF0000'}   # Red\n",
        "\n",
        "categories_labels = {1: 'Normal Operation',\n",
        "              2: 'Early Warning',\n",
        "              3: 'System Failure'}"
      ],
      "metadata": {
        "trusted": true,
        "id": "NnNCK0ib_FkH"
      },
      "outputs": [],
      "execution_count": 9
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 9.5. Filtering pairs Bacteria-Protein by significance to the risk category"
      ],
      "metadata": {
        "id": "JrjY2HKt_FkH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sample_contri = ECcontri_Uniprot_enriched.sample(n=5000, random_state=42) # row == zeile"
      ],
      "metadata": {
        "trusted": true,
        "id": "ZnCR4kbe_FkH"
      },
      "outputs": [],
      "execution_count": 12
    },
    {
      "cell_type": "code",
      "source": [
        "def prepare_base_matrix_stats(eccontri_df):\n",
        "    # Ensure 'idx' is the index if it's not already\n",
        "    if 'idx' not in eccontri_df.index.names:\n",
        "        eccontri_df = eccontri_df.set_index('idx')\n",
        "\n",
        "    # Group by 'Category', 'Genus', and 'protein_name', and sum 'norm_abund_contri'\n",
        "    base_matrix = eccontri_df.groupby(['Genus', 'protein_name', 'Category'], observed=True, as_index=True)['norm_abund_contri'].sum()\n",
        "\n",
        "    base_matrix = base_matrix.reset_index(drop=False)\n",
        "    return base_matrix"
      ],
      "metadata": {
        "trusted": true,
        "id": "G5iPJhl3_FkH"
      },
      "outputs": [],
      "execution_count": 13
    },
    {
      "cell_type": "code",
      "source": [
        "base_matrix = prepare_base_matrix_stats(sample_contri)"
      ],
      "metadata": {
        "trusted": true,
        "id": "D7q7GJG7_FkH"
      },
      "outputs": [],
      "execution_count": 14
    },
    {
      "cell_type": "code",
      "source": [
        "def perform_abundance_analysis(base_matrix, eccontri_df, alpha=0.05, n_jobs=-1):\n",
        "    \"\"\"\n",
        "    Performs statistical analysis on protein-genus pairs focusing on abundance patterns.\n",
        "\n",
        "    This function:\n",
        "    1. Identifies patterns based on category presence/absence AND abundance\n",
        "    2. Performs statistical tests where possible (pairs in multiple categories)\n",
        "    3. Identifies category-specific and abundant proteins\n",
        "\n",
        "    Parameters:  base_matrix :  DataFrame with columns 'Genus', 'protein_name', 'Category', 'norm_abund_contri'\n",
        "    alpha : float, default=0.05   Significance level for statistical tests\n",
        "    n_jobs : int, default=-1   Number of parallel jobs to run\n",
        "\n",
        "    Returns:\n",
        "    --------\n",
        "    tuple: (results_df, category_stats)\n",
        "        results_df: DataFrame with statistical results and patterns\n",
        "        category_stats: Dictionary with summary statistics\n",
        "    \"\"\"\n",
        "    # Start timer\n",
        "    start_time = time.time()\n",
        "\n",
        "    # Get unique categories and genus-protein pairs\n",
        "    categories = sorted(base_matrix[\"Category\"].unique())\n",
        "    print(f\"Found {len(categories)} categories: {categories}\")\n",
        "\n",
        "    pairs = base_matrix[['Genus', 'protein_name']].drop_duplicates()\n",
        "    print(f\"Total unique genus-protein pairs unique: {len(pairs)}\")\n",
        "\n",
        "    # Pre-compute statistics for all genus-protein-category combinations\n",
        "    print(\"Computing statistics for all genus-protein-category combinations...\")\n",
        "    grouped_stats = base_matrix.groupby(['Genus', 'protein_name', 'Category'])['norm_abund_contri'].agg([\n",
        "        ('mean', 'mean'),\n",
        "        ('median', 'median'),\n",
        "        ('std', 'std'),\n",
        "        ('count', 'count'),\n",
        "        ('sum', 'sum'),\n",
        "        ('values', lambda x: list(x))\n",
        "    ]).reset_index()\n",
        "\n",
        "    # Convert to dictionary for faster lookups and Create a list of tuples for keys and a list of dictionaries for values\n",
        "    keys = list(zip(grouped_stats['Genus'], grouped_stats['protein_name'], grouped_stats['Category'])) #generator\n",
        "    values = grouped_stats[['mean', 'median', 'std', 'count', 'sum', 'values']].to_dict('records')\n",
        "    grouped_dict = dict(zip(keys, values))\n",
        "\n",
        "    # Create a pivot table to easily see category presence\n",
        "    category_pivot = grouped_stats.pivot_table(\n",
        "        index=['Genus', 'protein_name'],\n",
        "        columns='Category',\n",
        "        values='count',\n",
        "        fill_value=0\n",
        "    ).reset_index()\n",
        "\n",
        "    # Add columns for category presence\n",
        "    for cat in categories:\n",
        "        category_pivot[f'present_in_cat{cat}'] = category_pivot[cat] > 0\n",
        "\n",
        "    def cliffs_delta(x, y):\n",
        "        \"\"\"\n",
        "        Calculate Cliff's Delta - a non-parametric effect size measure.\n",
        "        Returns value between -1 and 1, where:\n",
        "        -1: all values in y > values in x\n",
        "        +1: all values in x > values in y\n",
        "        0: distributions overlap completely\n",
        "        \"\"\"\n",
        "        # Count comparisons where x > y and y > x\n",
        "        greater = 0\n",
        "        less = 0\n",
        "\n",
        "        for i in x:\n",
        "            for j in y:\n",
        "                if i > j:\n",
        "                    greater += 1\n",
        "                elif i < j:\n",
        "                    less += 1\n",
        "\n",
        "        # Calculate delta\n",
        "        delta = (greater - less) / (len(x) * len(y))\n",
        "        return delta\n",
        "\n",
        "    # Define function to process each pair\n",
        "    def process_pair(row_data):\n",
        "        # Get category presence for this pair\n",
        "        genus = row_data['Genus']\n",
        "        protein = row_data['protein_name']\n",
        "\n",
        "        pivot_row = category_pivot[\n",
        "            (category_pivot['Genus'] == genus) &\n",
        "            (category_pivot['protein_name'] == protein)\n",
        "        ]\n",
        "\n",
        "        if pivot_row.empty:\n",
        "            return None\n",
        "\n",
        "        pivot_row = pivot_row.iloc[0]\n",
        "\n",
        "        # Check which categories have data\n",
        "        present_cats = []\n",
        "        for cat in categories:\n",
        "            if pivot_row[f'present_in_cat{cat}']:\n",
        "                present_cats.append(cat)\n",
        "\n",
        "        # Skip if no categories have data (shouldn't happen)\n",
        "        if not present_cats:\n",
        "            return None\n",
        "\n",
        "        # Compile stats for each category\n",
        "        cat_stats = {}\n",
        "        for cat in categories:\n",
        "            key = (genus, protein, cat)\n",
        "            if key in grouped_dict:\n",
        "                cat_stats[f\"mean_cat{cat}\"] = grouped_dict[key]['mean']\n",
        "                cat_stats[f\"median_cat{cat}\"] = grouped_dict[key]['median']\n",
        "                cat_stats[f\"std_cat{cat}\"] = grouped_dict[key]['std']\n",
        "                cat_stats[f\"count_cat{cat}\"] = grouped_dict[key]['count']\n",
        "                cat_stats[f\"sum_cat{cat}\"] = grouped_dict[key]['sum']\n",
        "            else:\n",
        "                cat_stats[f\"mean_cat{cat}\"] = np.nan\n",
        "                cat_stats[f\"median_cat{cat}\"] = np.nan\n",
        "                cat_stats[f\"std_cat{cat}\"] = np.nan\n",
        "                cat_stats[f\"count_cat{cat}\"] = np.nan\n",
        "                cat_stats[f\"sum_cat{cat}\"] = np.nan\n",
        "\n",
        "        # First determine presence pattern\n",
        "        pattern = \"unknown\"\n",
        "        category_str = ''.join(str(c) for c in present_cats)\n",
        "\n",
        "        if category_str == '123':\n",
        "            presence_pattern = 'all_categories'\n",
        "        elif category_str == '12':\n",
        "            presence_pattern = 'cat1_and_cat2'\n",
        "        elif category_str == '13':\n",
        "            presence_pattern = 'cat1_and_cat3'\n",
        "        elif category_str == '23':\n",
        "            presence_pattern = 'cat2_and_cat3'\n",
        "        elif category_str == '1':\n",
        "            presence_pattern = 'only_cat1'\n",
        "        elif category_str == '2':\n",
        "            presence_pattern = 'only_cat2'\n",
        "        elif category_str == '3':\n",
        "            presence_pattern = 'only_cat3'\n",
        "        else:\n",
        "            presence_pattern = 'unknown'\n",
        "\n",
        "        # Determine pattern based on statistical criteria:\n",
        "        pct_change_1to2 = np.nan  # Initialize with NaN\n",
        "        pct_change_2to3 = np.nan  # Initialize with NaN\n",
        "        # Initialize mean1, mean2, mean3\n",
        "        mean1 = np.nan\n",
        "        mean2 = np.nan\n",
        "        mean3 = np.nan\n",
        "        if presence_pattern == 'all_categories':\n",
        "            mean1, mean2, mean3 = cat_stats[\"mean_cat1\"], cat_stats[\"mean_cat2\"], cat_stats[\"mean_cat3\"]\n",
        "\n",
        "            # Calculate percent changes between categories\n",
        "            pct_change_1to2 = (mean2 - mean1) / mean1 if mean1 > 0 else np.inf\n",
        "            pct_change_2to3 = (mean3 - mean2) / mean2 if mean2 > 0 else np.inf\n",
        "\n",
        "            # Set minimum threshold for meaningful change (e.g., 20%)\n",
        "            significant_change_threshold = 0.1\n",
        "\n",
        "            # Classify patterns with statistical criteria\n",
        "            if pct_change_1to2 > significant_change_threshold and pct_change_2to3 > significant_change_threshold:\n",
        "                pattern = \"increasing_abundance\"\n",
        "            elif pct_change_1to2 < -significant_change_threshold and pct_change_2to3 < -significant_change_threshold:\n",
        "                pattern = \"decreasing_abundance\"\n",
        "            elif pct_change_1to2 > significant_change_threshold and pct_change_2to3 < -significant_change_threshold:\n",
        "                pattern = \"peak_at_cat2\"\n",
        "            elif pct_change_1to2 < -significant_change_threshold and pct_change_2to3 > significant_change_threshold:\n",
        "                pattern = \"valley_at_cat2\"\n",
        "            else:\n",
        "                pattern = \"mixed_abundance\"\n",
        "        else:\n",
        "            pattern = presence_pattern\n",
        "\n",
        "        print(f\"pct_change_2to3: {pct_change_2to3}, pct_change_1to2: {pct_change_1to2}\")\n",
        "        print(f\"Processing {genus}-{protein}, Present Categories: {present_cats}, category_str: {category_str}\")\n",
        "        print(f\"Mean values: {mean1}, {mean2}, {mean3}\")\n",
        "        print(f\"Percent changes: {pct_change_1to2}, {pct_change_2to3}\")\n",
        "        print(f\"Assigned pattern: {pattern}\")\n",
        "\n",
        "        # Calculate effect sizes between consecutive categories\n",
        "        effect_sizes = {}\n",
        "        for j in range(len(categories)-1):\n",
        "            cat1, cat2 = categories[j], categories[j+1]\n",
        "\n",
        "            if cat_stats[f\"count_cat{cat1}\"] > 0 and cat_stats[f\"count_cat{cat2}\"] > 0:\n",
        "                key1 = (genus, protein, cat1)\n",
        "                key2 = (genus, protein, cat2)\n",
        "\n",
        "                if key1 in grouped_dict and key2 in grouped_dict:\n",
        "                    values1 = grouped_dict[key1]['values']\n",
        "                    values2 = grouped_dict[key2]['values']\n",
        "\n",
        "                    if len(values1) > 0 and len(values2) > 0:\n",
        "                        # Calculate Cliff's delta\n",
        "                        effect_sizes[f\"effect_size_{cat1}_to_{cat2}\"] = cliffs_delta(values1, values2)\n",
        "                    else:\n",
        "                        effect_sizes[f\"effect_size_{cat1}_to_{cat2}\"] = np.nan\n",
        "                else:\n",
        "                    effect_sizes[f\"effect_size_{cat1}_to_{cat2}\"] = np.nan\n",
        "            else:\n",
        "                effect_sizes[f\"effect_size_{cat1}_to_{cat2}\"] = np.nan\n",
        "\n",
        "        # Statistical tests for pairs in multiple categories\n",
        "        # Initializing h_stat and p_val for the case when there is no sufficient value or/and are nans\n",
        "        h_stat = np.nan\n",
        "        p_val = np.nan\n",
        "        corr_value = 0.0\n",
        "\n",
        "        if len(present_cats) >= 2:\n",
        "            # Get values for each category\n",
        "            cat_values = []\n",
        "            sample_sizes = []\n",
        "            for cat in present_cats:\n",
        "                key = (genus, protein, cat)\n",
        "                if key in grouped_dict and len(grouped_dict[key]['values']) > 0:\n",
        "                    cat_values.append(grouped_dict[key]['values'])\n",
        "                    sample_sizes.append(len(grouped_dict[key]['values']))\n",
        "\n",
        "            # Check if we have adequate samples for testing, the h_stat was getting errors even with initialisation and exceptions, so we move it inside the block on purpose\n",
        "            if len(cat_values) >= 2 and min(sample_sizes) >= 3 and all(  # Minimum 3 samples per group\n",
        "                (genus, protein, cat) in grouped_dict and len(grouped_dict[(genus, protein, cat)]['values']) > 0\n",
        "                for cat in present_cats):\n",
        "\n",
        "                try:\n",
        "                    if len(cat_values) == 2:\n",
        "                        # Mann-Whitney U test for two categories\n",
        "                        # Check for ties and use exact method when appropriate\n",
        "                        if max(sample_sizes) > 20:  # Large sample approximation\n",
        "                            stat, p_val = mannwhitneyu(cat_values[0], cat_values[1], alternative='two-sided')\n",
        "                        else:\n",
        "                            # Use exact method for small samples\n",
        "                            stat, p_val = mannwhitneyu(cat_values[0], cat_values[1], alternative='two-sided', method='exact')\n",
        "                    else:\n",
        "                        # Kruskal-Wallis test for more than two categories\n",
        "                        h_stat, p_val = kruskal(*cat_values, nan_policy='omit')\n",
        "\n",
        "                    # Record the sample sizes for context\n",
        "                    effect_sizes['n_samples'] = sum(sample_sizes)\n",
        "                except Exception as e:\n",
        "                    print(f\"Statistical test failed: {e}\")\n",
        "                    h_stat, p_val = np.nan, np.nan\n",
        "\n",
        "            # Calculate correlation using all individual measurements\n",
        "            corr_value = 0.0\n",
        "            if len(present_cats) >= 2:\n",
        "                try:\n",
        "                    # Collect all individual measurements\n",
        "                    all_values = []\n",
        "                    all_categories = []\n",
        "\n",
        "                    for cat in present_cats:\n",
        "                        key = (genus, protein, cat)\n",
        "                        if key in grouped_dict and len(grouped_dict[key]['values']) > 0:\n",
        "                            all_values.extend(grouped_dict[key]['values'])\n",
        "                            all_categories.extend([cat] * len(grouped_dict[key]['values']))\n",
        "\n",
        "                    # Convert to numpy arrays\n",
        "                    all_values = np.array(all_values)\n",
        "                    all_categories = np.array(all_categories)\n",
        "\n",
        "                    # Use Spearman correlation (robust, non-parametric)\n",
        "                    from scipy.stats import spearmanr\n",
        "                    if len(all_values) >= 5:  # Minimum points for reliable correlation\n",
        "                        corr, p = spearmanr(all_categories, all_values)\n",
        "                        corr_value = corr\n",
        "                    else:\n",
        "                        # For few points, calculate trend direction\n",
        "                        trend = np.polyfit(all_categories, all_values, 1)[0]\n",
        "                        corr_value = np.sign(trend) * 0.5  # Scale to (-0.5, 0.5) for limited data\n",
        "                except Exception as e:\n",
        "                    print(f\"Correlation calculation error: {e}\")\n",
        "                    corr_value = 0.0\n",
        "\n",
        "        # Create result dictionary\n",
        "        result = {\n",
        "            'Genus': genus,\n",
        "            'protein_name': protein,\n",
        "            'h_statistic': h_stat,\n",
        "            'p_value': p_val,\n",
        "            'pattern': pattern,\n",
        "            'presence_pattern': presence_pattern,\n",
        "            'category_count': len(present_cats),\n",
        "            'categories': category_str,\n",
        "            'corr': corr_value,\n",
        "            **cat_stats,\n",
        "            **effect_sizes\n",
        "        }\n",
        "\n",
        "        return result\n",
        "\n",
        "    # Setup progress reporting\n",
        "    total_pairs = len(pairs)\n",
        "    start_time = time.time()\n",
        "    last_print_time = start_time\n",
        "\n",
        "    # Process in batches with reporting every 10 minutes\n",
        "    batch_size = 10000\n",
        "    results = []\n",
        "\n",
        "    for i in range(0, len(pairs), batch_size):\n",
        "        batch_end = min(i + batch_size, len(pairs))\n",
        "        batch_pairs = pairs.iloc[i:batch_end]\n",
        "\n",
        "        batch_results = Parallel(n_jobs=n_jobs)(\n",
        "            delayed(process_pair)(row)\n",
        "            for _, row in batch_pairs.iterrows())\n",
        "\n",
        "        results.extend([r for r in batch_results if r is not None])\n",
        "\n",
        "        # Print progress every 10 minutes\n",
        "        current_time = time.time()\n",
        "        if current_time - last_print_time >= 600:  # 600 seconds = 10 minutes\n",
        "            elapsed = current_time - start_time\n",
        "            progress = batch_end / total_pairs * 100\n",
        "            print(f\"Progress: {batch_end}/{total_pairs} pairs ({progress:.1f}%) - Elapsed time: {elapsed:.1f} seconds\")\n",
        "            last_print_time = current_time\n",
        "\n",
        "    # Remove None results\n",
        "    results = [r for r in results if r is not None]\n",
        "    print(f\"Generated {len(results)} valid results\")\n",
        "\n",
        "    # Create results DataFrame\n",
        "    if not results:\n",
        "        print(\"Warning: No valid statistical test results were produced.\")\n",
        "        return pd.DataFrame(), {}\n",
        "\n",
        "    results_df = pd.DataFrame(results)\n",
        "\n",
        "    # Perform multiple testing correction with FDR\n",
        "    if 'p_value' in results_df.columns:\n",
        "        valid_pvals = ~results_df['p_value'].isna()\n",
        "        if sum(valid_pvals) > 0:\n",
        "            # Create columns for adjusted p-values and significance\n",
        "            results_df['p_adjusted'] = np.nan\n",
        "            results_df['significant'] = False\n",
        "\n",
        "            # Apply Benjamini-Hochberg correction\n",
        "            from statsmodels.stats.multitest import multipletests\n",
        "            reject, pvals_corrected, _, _ = multipletests(\n",
        "                results_df.loc[valid_pvals, 'p_value'].values,\n",
        "                alpha=alpha,\n",
        "                method='fdr_bh'  # Benjamini-Hochberg FDR correction\n",
        "            )\n",
        "\n",
        "            # Store results\n",
        "            results_df.loc[valid_pvals, 'p_adjusted'] = pvals_corrected\n",
        "            results_df.loc[valid_pvals, 'significant'] = reject\n",
        "\n",
        "            print(f\"After FDR correction: {sum(reject)}/{sum(valid_pvals)} tests significant at alpha={alpha}\")\n",
        "\n",
        "    # Add total abundance column (sum across all categories)\n",
        "    results_df['total_abundance'] = 0\n",
        "    for cat in categories:\n",
        "        results_df['total_abundance'] += results_df[f'sum_cat{cat}'].fillna(0)\n",
        "\n",
        "    # Define safe_log2fc function\n",
        "    def safe_log2fc(value1, value2, min_value=1e-10):\n",
        "        \"\"\"\n",
        "        Calculate log2 fold change with safety checks for zero/NaN values.\n",
        "        \"\"\"\n",
        "        # Handle NaN values - keep as NaN rather than converting to 0\n",
        "        if pd.isna(value1) or pd.isna(value2):\n",
        "            return np.nan\n",
        "\n",
        "        # Handle zeros\n",
        "        v1 = max(float(value1), min_value)\n",
        "        v2 = max(float(value2), min_value)\n",
        "\n",
        "        # Calculate fold change\n",
        "        try:\n",
        "            return np.log2(v1 / v2)\n",
        "        except:\n",
        "            return np.nan\n",
        "\n",
        "    # Create masks for valid comparisons\n",
        "    mask_cat1_cat2 = (results_df['count_cat1'] > 0) & (results_df['count_cat2'] > 0)\n",
        "    mask_cat1_cat3 = (results_df['count_cat1'] > 0) & (results_df['count_cat3'] > 0)\n",
        "    mask_cat2_cat3 = (results_df['count_cat2'] > 0) & (results_df['count_cat3'] > 0)\n",
        "\n",
        "    # Initialize with NaN\n",
        "    results_df['log2fc_cat2_vs_cat1'] = np.nan\n",
        "    results_df['log2fc_cat3_vs_cat1'] = np.nan\n",
        "    results_df['log2fc_cat3_vs_cat2'] = np.nan\n",
        "\n",
        "    # Only calculate for valid pairs using vectorized operations\n",
        "    if sum(mask_cat1_cat2) > 0:\n",
        "        cat1_values = results_df.loc[mask_cat1_cat2, 'mean_cat1'].values\n",
        "        cat2_values = results_df.loc[mask_cat1_cat2, 'mean_cat2'].values\n",
        "        cat1_values = np.maximum(cat1_values, 1e-10)\n",
        "        cat2_values = np.maximum(cat2_values, 1e-10)\n",
        "        results_df.loc[mask_cat1_cat2, 'log2fc_cat2_vs_cat1'] = np.log2(cat2_values / cat1_values)\n",
        "\n",
        "    if sum(mask_cat1_cat3) > 0:\n",
        "        cat1_values = results_df.loc[mask_cat1_cat3, 'mean_cat1'].values\n",
        "        cat3_values = results_df.loc[mask_cat1_cat3, 'mean_cat3'].values\n",
        "        cat1_values = np.maximum(cat1_values, 1e-10)\n",
        "        cat3_values = np.maximum(cat3_values, 1e-10)\n",
        "        results_df.loc[mask_cat1_cat3, 'log2fc_cat3_vs_cat1'] = np.log2(cat3_values / cat1_values)\n",
        "\n",
        "    if sum(mask_cat2_cat3) > 0:\n",
        "        cat2_values = results_df.loc[mask_cat2_cat3, 'mean_cat2'].values\n",
        "        cat3_values = results_df.loc[mask_cat2_cat3, 'mean_cat3'].values\n",
        "        cat2_values = np.maximum(cat2_values, 1e-10)\n",
        "        cat3_values = np.maximum(cat3_values, 1e-10)\n",
        "        results_df.loc[mask_cat2_cat3, 'log2fc_cat3_vs_cat2'] = np.log2(cat3_values / cat2_values)\n",
        "\n",
        "    # Print a progress update\n",
        "    print(f\"Added log2 fold change calculations for {sum(mask_cat1_cat2)} cat1-cat2 pairs, {sum(mask_cat1_cat3)} cat1-cat3 pairs, {sum(mask_cat2_cat3)} cat2-cat3 pairs\")\n",
        "    # Create a dictionary with category statistics\n",
        "    category_stats = {\n",
        "        'total_pairs': len(results_df),\n",
        "        'tested_pairs': sum(~results_df['p_value'].isna()),\n",
        "        'significant_pairs': sum(results_df['significant']) if 'significant' in results_df.columns else 0\n",
        "    }\n",
        "\n",
        "    # Add pattern counts to stats\n",
        "    if 'pattern' in results_df.columns:\n",
        "        for pattern in results_df['pattern'].unique():\n",
        "            category_stats[f'pattern_{pattern}'] = sum(results_df['pattern'] == pattern)\n",
        "\n",
        "    results_df = results_df.reset_index(drop=False)\n",
        "\n",
        "    return results_df, category_stats"
      ],
      "metadata": {
        "trusted": true,
        "id": "dKqudUcZ_FkH"
      },
      "outputs": [],
      "execution_count": 109
    },
    {
      "cell_type": "code",
      "source": [
        "def process_in_resumable_chunks(base_matrix, output_dir, eccontri_df, chunk_size=50000):\n",
        "    \"\"\"\n",
        "    Process data in chunks and save intermediate results to allow resuming.\n",
        "\n",
        "    Parameters:\n",
        "    base_matrix : DataFrame to process\n",
        "    output_dir : Directory to save chunk results\n",
        "    eccontri_df : Original dataframe with idx information\n",
        "    chunk_size : Number of rows per chunk\n",
        "\n",
        "    Returns:\n",
        "    DataFrame with combined results\n",
        "    \"\"\"\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "    # Get total chunks needed\n",
        "    total_rows = len(base_matrix)\n",
        "    total_chunks = (total_rows + chunk_size - 1) // chunk_size\n",
        "\n",
        "    # Check for existing chunks\n",
        "    existing_chunks = [f for f in os.listdir(output_dir) if f.startswith('pattern_chunk_') and f.endswith('.csv')]\n",
        "    existing_nums = sorted([int(f.split('_')[-1].split('.')[0]) for f in existing_chunks])\n",
        "\n",
        "    print(f\"Found {len(existing_nums)} existing chunks out of {total_chunks} total needed\")\n",
        "\n",
        "    # Process remaining chunks\n",
        "    for chunk_num in range(1, total_chunks + 1):\n",
        "        # Skip if already processed\n",
        "        if chunk_num in existing_nums:\n",
        "            print(f\"Skipping chunk {chunk_num}/{total_chunks} - already processed\")\n",
        "            continue\n",
        "\n",
        "        # Process this chunk\n",
        "        start_idx = (chunk_num - 1) * chunk_size\n",
        "        end_idx = min(start_idx + chunk_size, total_rows)\n",
        "\n",
        "        print(f\"Processing chunk {chunk_num}/{total_chunks}: rows {start_idx} to {end_idx}\")\n",
        "        chunk_data = base_matrix.iloc[start_idx:end_idx]\n",
        "\n",
        "        # Run analysis on this chunk\n",
        "        chunk_results, _ = perform_abundance_analysis(chunk_data,eccontri_df, alpha=0.05, n_jobs=-1)\n",
        "\n",
        "        # Save results\n",
        "        output_file = os.path.join(output_dir, f\"pattern_chunk_{chunk_num}.parquet\")\n",
        "        chunk_results.to_parquet(output_file, index=False)\n",
        "        print(f\"Saved results for chunk {chunk_num} to {output_file}\")\n",
        "\n",
        "    # Combine all chunks once complete\n",
        "    print(\"Combining all chunks...\")\n",
        "    all_chunks = []\n",
        "    for chunk_num in range(1, total_chunks + 1):\n",
        "        file_path = os.path.join(output_dir, f\"pattern_chunk_{chunk_num}.parquet\")\n",
        "        if os.path.exists(file_path):\n",
        "            chunk_df = pd.read_parquet(file_path)\n",
        "            all_chunks.append(chunk_df)\n",
        "\n",
        "    if all_chunks:\n",
        "        combined_results = pd.concat(all_chunks, ignore_index=True)\n",
        "        print(f\"Combined {len(all_chunks)} chunks into final result with {len(combined_results)} rows\")\n",
        "        return combined_results\n",
        "    else:\n",
        "        print(\"No chunks were processed\")\n",
        "        return pd.DataFrame()"
      ],
      "metadata": {
        "trusted": true,
        "id": "kZzxw1ej_FkI"
      },
      "outputs": [],
      "execution_count": 110
    },
    {
      "cell_type": "code",
      "source": [
        "# Reshape data for analysis and calculate abundance matrix\n",
        "print(\"calculating abundancy pattern...\")\n",
        "output_dir = \"pattern_analysis_results\"\n",
        "pattern_df = process_in_resumable_chunks(base_matrix,  output_dir, sample_contri, chunk_size=50000)"
      ],
      "metadata": {
        "trusted": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GcOM0hs0_FkI",
        "outputId": "2bab1148-334d-440d-ed33-e2e597f07c31"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "calculating abundancy pattern...\n",
            "Found 0 existing chunks out of 1 total needed\n",
            "Processing chunk 1/1: rows 0 to 4875\n",
            "Found 3 categories: [np.int64(1), np.int64(2), np.int64(3)]\n",
            "Total unique genus-protein pairs unique: 4668\n",
            "Computing statistics for all genus-protein-category combinations...\n",
            "Generated 4668 valid results\n",
            "Added log2 fold change calculations for 74 cat1-cat2 pairs, 52 cat1-cat3 pairs, 91 cat2-cat3 pairs\n",
            "Saved results for chunk 1 to pattern_analysis_results/pattern_chunk_1.parquet\n",
            "Combining all chunks...\n",
            "Combined 1 chunks into final result with 4668 rows\n"
          ]
        }
      ],
      "execution_count": 111
    },
    {
      "cell_type": "markdown",
      "source": [
        "# To put on the main function after the chunking"
      ],
      "metadata": {
        "id": "4wZfNnIPcL3o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sample_contri.head()"
      ],
      "metadata": {
        "id": "oeq1ArdPiyYH",
        "outputId": "b10517c9-fa95-4ae4-d6a8-ddb21e4123d2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 551
        }
      },
      "execution_count": 117,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       idx    Sites           Genus  abund_raw  rel_abund_raw  \\\n",
              "0   502012  site_27          Afipia   267830.0      28.021013   \n",
              "1   562124  site_30   Anoxybacillus    32682.0       4.590262   \n",
              "2   472739  site_25    Sphingomonas     1121.0       0.168339   \n",
              "3   133617   site_7  Staphylococcus     7736.0       0.878632   \n",
              "4  1136015  site_63  Bradyrhizobium     4161.0       0.613867   \n",
              "\n",
              "   genome_EC_count  abund_contri  rel_abund_contri  norm_abund_contri  \\\n",
              "0                2      535660.0         56.042026           0.353589   \n",
              "1                2       65364.0          9.180525           0.054352   \n",
              "2                1        1121.0          0.168339           0.001973   \n",
              "3                1        7736.0          0.878632           0.940890   \n",
              "4                8       33288.0          4.910938           0.011502   \n",
              "\n",
              "                                        protein_name  ...  \\\n",
              "0                             prephenate dehydratase  ...   \n",
              "1  corrinoid adenosyltransferase alamin adenosylt...  ...   \n",
              "2                    phosphoenolpyruvate carboxylase  ...   \n",
              "3  cystathionine gamma-lyase and homocysteine for...  ...   \n",
              "4                acetolactate-synthase small subunit  ...   \n",
              "\n",
              "                                            pathways  \\\n",
              "0  Phenylalanine, tyrosine and tryptophan biosynt...   \n",
              "1  Arginine biosynthesis; Arginine and proline me...   \n",
              "2  Glycolysis / Gluconeogenesis; Citrate cycle (T...   \n",
              "3  Diterpenoid biosynthesis; Biosynthesis of seco...   \n",
              "4  Valine, leucine and isoleucine biosynthesis; B...   \n",
              "\n",
              "                                           hierarchy  \\\n",
              "0  Amino acid metabolism > Phenylalanine, tyrosin...   \n",
              "1  Cardiovascular disease > Viral myocarditis; Ne...   \n",
              "2  Energy metabolism > Carbon fixation by Calvin ...   \n",
              "3  Metabolism of terpenoids and polyketides > Dit...   \n",
              "4  Metabolism of cofactors and vitamins > Pantoth...   \n",
              "\n",
              "                                     metals_involved  \\\n",
              "0  potassium; phosphate; iron; sulfur; hydrogen; ...   \n",
              "1  Mo; iron; zinc; hydrogen; selenium; nitrate; n...   \n",
              "2  Mo; iron; sulfur; hydrogen; sodium; selenium; ...   \n",
              "3                                           selenium   \n",
              "4                Al3+; Mo; iron; selenium; phosphate   \n",
              "\n",
              "                  metals_consolidated  \\\n",
              "0  K; phosphate; H; Mo; Mn; Se; Fe; S   \n",
              "1        H; Zn; Mo; NO3-; Se; Fe; NO2   \n",
              "2     H; phosphate; Mo; Se; Fe; S; Na   \n",
              "3                                  Se   \n",
              "4         phosphate; Mo; Se; Fe; Al3+   \n",
              "\n",
              "                                corrosion_mechanisms  \\\n",
              "0  h2_consumption; acid_production; sulfur_metabo...   \n",
              "1                         h2_consumption; direct_eet   \n",
              "2                    h2_consumption; acid_production   \n",
              "3                                               None   \n",
              "4                                    acid_production   \n",
              "\n",
              "  corrosion_relevance_score corrosion_relevance has_metal Category  \\\n",
              "0                      20.0                high      True        2   \n",
              "1                      14.5                high      True        2   \n",
              "2                      15.5                high      True        2   \n",
              "3                       1.5                 low      True        2   \n",
              "4                      10.5                high      True        2   \n",
              "\n",
              "   was_uncharacterized  \n",
              "0                False  \n",
              "1                False  \n",
              "2                False  \n",
              "3                False  \n",
              "4                False  \n",
              "\n",
              "[5 rows x 24 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-98632cf0-f560-4088-bc67-412aad61bf47\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>idx</th>\n",
              "      <th>Sites</th>\n",
              "      <th>Genus</th>\n",
              "      <th>abund_raw</th>\n",
              "      <th>rel_abund_raw</th>\n",
              "      <th>genome_EC_count</th>\n",
              "      <th>abund_contri</th>\n",
              "      <th>rel_abund_contri</th>\n",
              "      <th>norm_abund_contri</th>\n",
              "      <th>protein_name</th>\n",
              "      <th>...</th>\n",
              "      <th>pathways</th>\n",
              "      <th>hierarchy</th>\n",
              "      <th>metals_involved</th>\n",
              "      <th>metals_consolidated</th>\n",
              "      <th>corrosion_mechanisms</th>\n",
              "      <th>corrosion_relevance_score</th>\n",
              "      <th>corrosion_relevance</th>\n",
              "      <th>has_metal</th>\n",
              "      <th>Category</th>\n",
              "      <th>was_uncharacterized</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>502012</td>\n",
              "      <td>site_27</td>\n",
              "      <td>Afipia</td>\n",
              "      <td>267830.0</td>\n",
              "      <td>28.021013</td>\n",
              "      <td>2</td>\n",
              "      <td>535660.0</td>\n",
              "      <td>56.042026</td>\n",
              "      <td>0.353589</td>\n",
              "      <td>prephenate dehydratase</td>\n",
              "      <td>...</td>\n",
              "      <td>Phenylalanine, tyrosine and tryptophan biosynt...</td>\n",
              "      <td>Amino acid metabolism &gt; Phenylalanine, tyrosin...</td>\n",
              "      <td>potassium; phosphate; iron; sulfur; hydrogen; ...</td>\n",
              "      <td>K; phosphate; H; Mo; Mn; Se; Fe; S</td>\n",
              "      <td>h2_consumption; acid_production; sulfur_metabo...</td>\n",
              "      <td>20.0</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>562124</td>\n",
              "      <td>site_30</td>\n",
              "      <td>Anoxybacillus</td>\n",
              "      <td>32682.0</td>\n",
              "      <td>4.590262</td>\n",
              "      <td>2</td>\n",
              "      <td>65364.0</td>\n",
              "      <td>9.180525</td>\n",
              "      <td>0.054352</td>\n",
              "      <td>corrinoid adenosyltransferase alamin adenosylt...</td>\n",
              "      <td>...</td>\n",
              "      <td>Arginine biosynthesis; Arginine and proline me...</td>\n",
              "      <td>Cardiovascular disease &gt; Viral myocarditis; Ne...</td>\n",
              "      <td>Mo; iron; zinc; hydrogen; selenium; nitrate; n...</td>\n",
              "      <td>H; Zn; Mo; NO3-; Se; Fe; NO2</td>\n",
              "      <td>h2_consumption; direct_eet</td>\n",
              "      <td>14.5</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>472739</td>\n",
              "      <td>site_25</td>\n",
              "      <td>Sphingomonas</td>\n",
              "      <td>1121.0</td>\n",
              "      <td>0.168339</td>\n",
              "      <td>1</td>\n",
              "      <td>1121.0</td>\n",
              "      <td>0.168339</td>\n",
              "      <td>0.001973</td>\n",
              "      <td>phosphoenolpyruvate carboxylase</td>\n",
              "      <td>...</td>\n",
              "      <td>Glycolysis / Gluconeogenesis; Citrate cycle (T...</td>\n",
              "      <td>Energy metabolism &gt; Carbon fixation by Calvin ...</td>\n",
              "      <td>Mo; iron; sulfur; hydrogen; sodium; selenium; ...</td>\n",
              "      <td>H; phosphate; Mo; Se; Fe; S; Na</td>\n",
              "      <td>h2_consumption; acid_production</td>\n",
              "      <td>15.5</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>133617</td>\n",
              "      <td>site_7</td>\n",
              "      <td>Staphylococcus</td>\n",
              "      <td>7736.0</td>\n",
              "      <td>0.878632</td>\n",
              "      <td>1</td>\n",
              "      <td>7736.0</td>\n",
              "      <td>0.878632</td>\n",
              "      <td>0.940890</td>\n",
              "      <td>cystathionine gamma-lyase and homocysteine for...</td>\n",
              "      <td>...</td>\n",
              "      <td>Diterpenoid biosynthesis; Biosynthesis of seco...</td>\n",
              "      <td>Metabolism of terpenoids and polyketides &gt; Dit...</td>\n",
              "      <td>selenium</td>\n",
              "      <td>Se</td>\n",
              "      <td>None</td>\n",
              "      <td>1.5</td>\n",
              "      <td>low</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1136015</td>\n",
              "      <td>site_63</td>\n",
              "      <td>Bradyrhizobium</td>\n",
              "      <td>4161.0</td>\n",
              "      <td>0.613867</td>\n",
              "      <td>8</td>\n",
              "      <td>33288.0</td>\n",
              "      <td>4.910938</td>\n",
              "      <td>0.011502</td>\n",
              "      <td>acetolactate-synthase small subunit</td>\n",
              "      <td>...</td>\n",
              "      <td>Valine, leucine and isoleucine biosynthesis; B...</td>\n",
              "      <td>Metabolism of cofactors and vitamins &gt; Pantoth...</td>\n",
              "      <td>Al3+; Mo; iron; selenium; phosphate</td>\n",
              "      <td>phosphate; Mo; Se; Fe; Al3+</td>\n",
              "      <td>acid_production</td>\n",
              "      <td>10.5</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 24 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-98632cf0-f560-4088-bc67-412aad61bf47')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-98632cf0-f560-4088-bc67-412aad61bf47 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-98632cf0-f560-4088-bc67-412aad61bf47');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-bae9d700-ae36-4f69-b552-a5f7fde26333\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-bae9d700-ae36-4f69-b552-a5f7fde26333')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-bae9d700-ae36-4f69-b552-a5f7fde26333 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "sample_contri"
            }
          },
          "metadata": {},
          "execution_count": 117
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sample_contri  = sample_contri.drop(columns=[\"level_0\", \"index\"], axis=1)"
      ],
      "metadata": {
        "id": "oabc-Ypjibzn"
      },
      "execution_count": 116,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sample_contri = sample_contri.reset_index(drop=False)\n",
        "# Create mapping of protein-genus pairs to their sites\n",
        "protein_genus_sites = {}\n",
        "site_groups = sample_contri.groupby(['Genus', 'protein_name'])['Sites'].apply(list)\n",
        "for (genus, protein), sites in site_groups.items():\n",
        "    protein_genus_sites[(genus, protein)] = list(set(sites))\n",
        "sample_contri_idx = sample_contri.groupby(['Genus', 'protein_name'])['idx'].first().reset_index()\n",
        "# Perform a simple merge\n",
        "pattern_df = pattern_df.merge(\n",
        "        sample_contri_idx,\n",
        "        on=['Genus', 'protein_name'],\n",
        "        how='left'\n",
        ")\n",
        "# Keep idx_y and drop idx_x (or vice versa depending on which one you want)\n",
        "# Check if 'idx' already exists in pattern_df\n",
        "if 'idx' in pattern_df.columns:\n",
        "    # Rename the existing 'idx' column to avoid conflicts\n",
        "    pattern_df = pattern_df.rename(columns={'idx': 'idx_x'})\n",
        "\n",
        "# Keep idx from sample_contri_idx and drop the other if it exists\n",
        "pattern_df['idx'] = pattern_df['idx'] if 'idx' in pattern_df.columns else pattern_df['idx_x'] # Assign to 'idx' directly if it exists, otherwise use 'idx_x'\n",
        "pattern_df = pattern_df.drop(['idx_x'], axis=1, errors='ignore') # Drop 'idx_x' if it exists, ignore if not\n",
        "pattern_df = pattern_df.set_index(\"idx\")\n",
        "pattern_df = pattern_df.reset_index()\n",
        "pattern_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 360
        },
        "id": "hKLH7zW4ZTn9",
        "outputId": "4aef040b-5fc9-4642-8511-05e2ecf4c82a"
      },
      "execution_count": 119,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       idx  index           Genus  \\\n",
              "0   458692      0  Acetobacterium   \n",
              "1  1415491      1  Acetobacterium   \n",
              "2   956935      2  Acetobacterium   \n",
              "3  1291065      3  Acetobacterium   \n",
              "4    15480      4  Acetobacterium   \n",
              "\n",
              "                                        protein_name  h_statistic  p_value  \\\n",
              "0                 23s rrna )-methyltransferase rlmcd          NaN      NaN   \n",
              "1                   23s rrna-methyltransferase; tlya          NaN      NaN   \n",
              "2                                3-oxoacyl-reductase          NaN      NaN   \n",
              "3  4-carboxymuconolactone decarboxylase; gamma-4-...          NaN      NaN   \n",
              "4                             adenylosuccinate lyase          NaN      NaN   \n",
              "\n",
              "     pattern presence_pattern  category_count categories  ...  std_cat3  \\\n",
              "0  only_cat2        only_cat2               1          2  ...       NaN   \n",
              "1  only_cat2        only_cat2               1          2  ...       NaN   \n",
              "2  only_cat3        only_cat3               1          3  ...       NaN   \n",
              "3  only_cat1        only_cat1               1          1  ...       NaN   \n",
              "4  only_cat3        only_cat3               1          3  ...       NaN   \n",
              "\n",
              "   count_cat3  sum_cat3  effect_size_1_to_2  effect_size_2_to_3  \\\n",
              "0         NaN       NaN                 NaN                 NaN   \n",
              "1         NaN       NaN                 NaN                 NaN   \n",
              "2         1.0  0.037237                 NaN                 NaN   \n",
              "3         NaN       NaN                 NaN                 NaN   \n",
              "4         1.0  0.005887                 NaN                 NaN   \n",
              "\n",
              "   total_abundance  log2fc_cat2_vs_cat1  log2fc_cat3_vs_cat1  \\\n",
              "0         0.001766                  NaN                  NaN   \n",
              "1         0.008023                  NaN                  NaN   \n",
              "2         0.037237                  NaN                  NaN   \n",
              "3         0.000577                  NaN                  NaN   \n",
              "4         0.005887                  NaN                  NaN   \n",
              "\n",
              "   log2fc_cat3_vs_cat2    idx_y  \n",
              "0                  NaN   458692  \n",
              "1                  NaN  1415491  \n",
              "2                  NaN   956935  \n",
              "3                  NaN  1291065  \n",
              "4                  NaN    15480  \n",
              "\n",
              "[5 rows x 33 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-8aa4e986-0979-4e46-b72c-3ed63281e811\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>idx</th>\n",
              "      <th>index</th>\n",
              "      <th>Genus</th>\n",
              "      <th>protein_name</th>\n",
              "      <th>h_statistic</th>\n",
              "      <th>p_value</th>\n",
              "      <th>pattern</th>\n",
              "      <th>presence_pattern</th>\n",
              "      <th>category_count</th>\n",
              "      <th>categories</th>\n",
              "      <th>...</th>\n",
              "      <th>std_cat3</th>\n",
              "      <th>count_cat3</th>\n",
              "      <th>sum_cat3</th>\n",
              "      <th>effect_size_1_to_2</th>\n",
              "      <th>effect_size_2_to_3</th>\n",
              "      <th>total_abundance</th>\n",
              "      <th>log2fc_cat2_vs_cat1</th>\n",
              "      <th>log2fc_cat3_vs_cat1</th>\n",
              "      <th>log2fc_cat3_vs_cat2</th>\n",
              "      <th>idx_y</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>458692</td>\n",
              "      <td>0</td>\n",
              "      <td>Acetobacterium</td>\n",
              "      <td>23s rrna )-methyltransferase rlmcd</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>only_cat2</td>\n",
              "      <td>only_cat2</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.001766</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>458692</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1415491</td>\n",
              "      <td>1</td>\n",
              "      <td>Acetobacterium</td>\n",
              "      <td>23s rrna-methyltransferase; tlya</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>only_cat2</td>\n",
              "      <td>only_cat2</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.008023</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1415491</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>956935</td>\n",
              "      <td>2</td>\n",
              "      <td>Acetobacterium</td>\n",
              "      <td>3-oxoacyl-reductase</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>only_cat3</td>\n",
              "      <td>only_cat3</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.037237</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.037237</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>956935</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1291065</td>\n",
              "      <td>3</td>\n",
              "      <td>Acetobacterium</td>\n",
              "      <td>4-carboxymuconolactone decarboxylase; gamma-4-...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>only_cat1</td>\n",
              "      <td>only_cat1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.000577</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1291065</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>15480</td>\n",
              "      <td>4</td>\n",
              "      <td>Acetobacterium</td>\n",
              "      <td>adenylosuccinate lyase</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>only_cat3</td>\n",
              "      <td>only_cat3</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.005887</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.005887</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>15480</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 33 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-8aa4e986-0979-4e46-b72c-3ed63281e811')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-8aa4e986-0979-4e46-b72c-3ed63281e811 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-8aa4e986-0979-4e46-b72c-3ed63281e811');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-965f0177-154b-4825-bd43-c80b7769c09e\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-965f0177-154b-4825-bd43-c80b7769c09e')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-965f0177-154b-4825-bd43-c80b7769c09e button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "pattern_df"
            }
          },
          "metadata": {},
          "execution_count": 119
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pattern_df = pattern_df.drop(columns=\"index\", axis=1)"
      ],
      "metadata": {
        "id": "Zp7fjpB1RFQx"
      },
      "execution_count": 120,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "plt.figure(figsize=(12, 6))\n",
        "plt.hist(pattern_df['categories'].dropna(), bins=20)\n",
        "plt.tick_params(axis='both', which='major', labelsize=10)\n",
        "plt.xlabel('categories')\n",
        "plt.ylabel('Frequency')\n",
        "plt.title('Distribution of categories')\n",
        "plt.xticks(rotation=30, ha='right')\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ],
      "metadata": {
        "trusted": true,
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 329
        },
        "id": "LY0Jjc0m_FkI",
        "outputId": "a0f7377e-483b-479d-d5f1-9abf2e7df82f"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1200x600 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAJOCAYAAABm7rQwAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAWT5JREFUeJzt3Xm4VWXdP/73YTiHGZyYEgFRcUQckVDUUBFJM7U0JySUUswUS8JSAXvEMFFLU/ObUg459DiUU+A8YTnhgIlCKBaTI0dQGffvD3/sxyM4Ia6D8Hpd176us+773mt91na5Oby5170qSqVSKQAAAABQoDq1XQAAAAAAax6hFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFACw0g0bNiwVFRWFHGu33XbLbrvtVt6+7777UlFRkb/85S+FHP+oo45Khw4dCjnWipo7d26OPvrotG7dOhUVFTnxxBNru6RV0pgxY1JRUZGXX365tksBgDWCUAoA+ERL/6K+9NWgQYO0bds2vXv3zm9+85u88847K+U406dPz7BhwzJhwoSVsr+VaVWu7bM466yzMmbMmBx77LG58sorc8QRRxRy3GuuuSbnn39+IccCAL56KkqlUqm2iwAAVl1jxoxJ//79M2LEiHTs2DELFy7MzJkzc99992XcuHHZYIMN8te//jVdunQpv2fRokVZtGhRGjRo8JmP8/jjj2eHHXbIFVdckaOOOuozv2/BggVJksrKyiQfzJTafffdc8MNN+Sggw76zPtZ0doWLlyYJUuWpKqqaqUc68uw0047pV69ennooYcKPe43v/nNPPfcc1+ZmUeLFy/OwoULU1VVVdhMPwBYk9Wr7QIAgK+GPn36ZPvtty9vDx06NPfcc0+++c1vZr/99su//vWvNGzYMElSr1691Kv35f6a8e6776ZRo0blMKq21K9fv1aP/1nMnj07m2++eW2XscqaN29eGjdunLp166Zu3bq1XQ4ArDHcvgcArLBvfOMbOe200/LKK6/kqquuKrcvb02pcePGZeedd06LFi3SpEmTdO7cOaeeemqSD2Y37bDDDkmS/v37l28VHDNmTJIP1o3acsst88QTT6Rnz55p1KhR+b0fXVNqqcWLF+fUU09N69at07hx4+y333559dVXa4zp0KHDcmdlfXifn1bb8taUmjdvXk4++eS0a9cuVVVV6dy5c37961/noxPUKyoqcvzxx+fmm2/OlltumaqqqmyxxRa58847l/+Bf8Ts2bMzYMCAtGrVKg0aNMjWW2+dP/7xj+X+petrTZ06Nbfddlu59k+buXTVVVdlxx13TKNGjbLWWmulZ8+eGTt2bLn/lltuSd++fdO2bdtUVVWlU6dOOfPMM7N48eIan+Ftt92WV155pXzcD39O8+fPzxlnnJGNNtooVVVVadeuXU455ZTMnz+/Ri3vvfdeTjjhhKy77rpp2rRp9ttvv/z3v/9NRUVFhg0bVmPsU089lT59+qRZs2Zp0qRJevXqlUcffbTGmKW3o95///057rjj0rJly6y//vo1+j76+dxxxx3ZZZdd0rhx4zRt2jR9+/bNxIkTa4yZOXNm+vfvn/XXXz9VVVVp06ZNvvWtb31lZokBQG0wUwoA+EKOOOKInHrqqRk7dmyOOeaY5Y6ZOHFivvnNb6ZLly4ZMWJEqqqqMnny5Dz88MNJks022ywjRozI6aefnoEDB2aXXXZJknz9618v7+ONN95Inz59csghh+Twww9Pq1atPrGu//mf/0lFRUWGDBmS2bNn5/zzz88ee+yRCRMmlGd0fRafpbYPK5VK2W+//XLvvfdmwIAB6dq1a/7+97/npz/9af773//mvPPOqzH+oYceyo033pjjjjsuTZs2zW9+85sceOCBmTZtWtZZZ52Preu9997LbrvtlsmTJ+f4449Px44dc8MNN+Soo47K22+/nR//+MfZbLPNcuWVV+akk07K+uuvn5NPPjlJst56633sfocPH55hw4bl61//ekaMGJHKysr84x//yD333JO99toryQfhTZMmTTJ48OA0adIk99xzT04//fRUV1fnnHPOSZL8/Oc/z5w5c/Kf//ynfM5NmjRJkixZsiT77bdfHnrooQwcODCbbbZZnn322Zx33nl58cUXc/PNN5frOeqoo3L99dfniCOOyE477ZT7778/ffv2XabuiRMnZpdddkmzZs1yyimnpH79+rn00kuz22675f7770+3bt1qjD/uuOOy3nrr5fTTT8+8efM+9vO48sor069fv/Tu3Tu/+tWv8u677+biiy/OzjvvnKeeeqoctB144IGZOHFifvSjH6VDhw6ZPXt2xo0bl2nTpq3yC+EDQK0pAQB8giuuuKKUpPTYY4997JjmzZuXttlmm/L2GWecUfrwrxnnnXdeKUnptdde+9h9PPbYY6UkpSuuuGKZvl133bWUpHTJJZcst2/XXXctb997772lJKWvfe1rperq6nL79ddfX0pSuuCCC8pt7du3L/Xr1+9T9/lJtfXr16/Uvn378vbNN99cSlL65S9/WWPcQQcdVKqoqChNnjy53JakVFlZWaPt6aefLiUp/fa3v13mWB92/vnnl5KUrrrqqnLbggULSt27dy81adKkxrm3b9++1Ldv30/cX6lUKr300kulOnXqlL797W+XFi9eXKNvyZIl5Z/ffffdZd77gx/8oNSoUaPS+++/X27r27dvjc9mqSuvvLJUp06d0oMPPlij/ZJLLiklKT388MOlUqlUeuKJJ0pJSieeeGKNcUcddVQpSemMM84ot+2///6lysrK0pQpU8pt06dPLzVt2rTUs2fPctvS63nnnXcuLVq0qMZ+l/ZNnTq1VCqVSu+8806pRYsWpWOOOabGuJkzZ5aaN29ebn/rrbdKSUrnnHPOMucKAHw8t+8BAF9YkyZNPvEpfC1atEjywW1fS5YsWaFjVFVVpX///p95/JFHHpmmTZuWtw866KC0adMmt99++wod/7O6/fbbU7du3Zxwwgk12k8++eSUSqXccccdNdr32GOPdOrUqbzdpUuXNGvWLP/+978/9TitW7fO9773vXJb/fr1c8IJJ2Tu3Lm5//77P3ftN998c5YsWZLTTz89derU/DXxw7djfnim2TvvvJPXX389u+yyS95999288MILn3qcG264IZtttlk23XTTvP766+XXN77xjSTJvffemyTl2xiPO+64Gu//0Y9+VGN78eLFGTt2bPbff/9suOGG5fY2bdrk0EMPzUMPPZTq6uoa7znmmGM+df2ocePG5e233873vve9GnXWrVs33bp1K9fZsGHDVFZW5r777stbb731qecPAHxAKAUAfGFz586tEQB91MEHH5wePXrk6KOPTqtWrXLIIYfk+uuv/1wB1de+9rXPtaj5xhtvXGO7oqIiG2200Ze+xs8rr7yStm3bLvN5bLbZZuX+D9tggw2W2cdaa631qeHGK6+8ko033niZ8OjjjvNZTJkyJXXq1PnURdEnTpyYb3/722nevHmaNWuW9dZbL4cffniSZM6cOZ96nJdeeikTJ07MeuutV+O1ySabJPlgrayl51CnTp107Nixxvs32mijGtuvvfZa3n333XTu3HmZY2222WZZsmTJMuuJfXSfH1dn8sHaaR+tdezYseU6q6qq8qtf/Sp33HFHWrVqlZ49e2bUqFGZOXPmpx4DANZk1pQCAL6Q//znP5kzZ84yQcGHNWzYMA888EDuvffe3Hbbbbnzzjtz3XXX5Rvf+EbGjh37mZ549nnWgfqsProY+1KLFy8u7ClsH3ec0kcWRV9VvP3229l1113TrFmzjBgxIp06dUqDBg3y5JNPZsiQIZ8paFyyZEm22mqrjB49ern97dq1W9llL+OzXE9Lz+XKK69M69atl+n/8BMmTzzxxOy77765+eab8/e//z2nnXZaRo4cmXvuuSfbbLPNyiscAFYjQikA4Au58sorkyS9e/f+xHF16tRJr1690qtXr4wePTpnnXVWfv7zn+fee+/NHnvs8bEB0YpaOstlqVKplMmTJ6dLly7ltrXWWitvv/32Mu995ZVXatwG9nlqa9++fe6666688847NWZLLb2trX379p95X592nGeeeSZLliypMVvqixynU6dOWbJkSZ5//vl07dp1uWPuu+++vPHGG7nxxhvTs2fPcvvUqVOXGftxn1unTp3y9NNPp1evXp/42bZv3z5LlizJ1KlTa8x8mzx5co1x6623Xho1apRJkyYts48XXnghderUWaGga+ltlS1btswee+zxmcaffPLJOfnkk/PSSy+la9euOffcc2s8mRIA+D9u3wMAVtg999yTM888Mx07dsxhhx32sePefPPNZdqWhh7z589PkjRu3DhJlhsSrYg//elPNda5+stf/pIZM2akT58+5bZOnTrl0UcfzYIFC8ptt9566zK3en2e2vbZZ58sXrw4F154YY328847LxUVFTWO/0Xss88+mTlzZq677rpy26JFi/Lb3/42TZo0ya677vq597n//vunTp06GTFixDIznpbO3Fo6s+vDM7kWLFiQ3/3ud8vsr3Hjxsu9ne+73/1u/vvf/+ayyy5bpu+9994rPw1vadD50X3/9re/rbFdt27d7LXXXrnllltq3J45a9asXHPNNdl5553TrFmzjz3vj9O7d+80a9YsZ511VhYuXLhM/2uvvZYkeffdd/P+++/X6OvUqVOaNm1avr4BgGWZKQUAfCZ33HFHXnjhhSxatCizZs3KPffck3HjxqV9+/b561//mgYNGnzse0eMGJEHHnggffv2Tfv27TN79uz87ne/y/rrr5+dd945yQd/iW/RokUuueSSNG3aNI0bN063bt0+09o/y7P22mtn5513Tv/+/TNr1qycf/752WijjXLMMceUxxx99NH5y1/+kr333jvf/e53M2XKlFx11VU1Fh7/vLXtu+++2X333fPzn/88L7/8crbeeuuMHTs2t9xyS0488cRl9r2iBg4cmEsvvTRHHXVUnnjiiXTo0CF/+ctf8vDDD+f888//xDW+Ps5GG22Un//85znzzDOzyy675IADDkhVVVUee+yxtG3bNiNHjszXv/71rLXWWunXr19OOOGEVFRU5Morr1zu7YbbbbddrrvuugwePDg77LBDmjRpkn333TdHHHFErr/++vzwhz/Mvffemx49emTx4sV54YUXcv311+fvf/97tt9++2y33XY58MADc/755+eNN97ITjvtlPvvvz8vvvhikpozsX75y19m3Lhx2XnnnXPcccelXr16ufTSSzN//vyMGjVqhT7jZs2a5eKLL84RRxyRbbfdNoccckjWW2+9TJs2Lbfddlt69OiRCy+8MC+++GJ69eqV7373u9l8881Tr1693HTTTZk1a1YOOeSQFTo2AKwRavPRfwDAqu+KK64oJSm/KisrS61bty7tueeepQsuuKBUXV29zHvOOOOM0od/zbj77rtL3/rWt0pt27YtVVZWltq2bVv63ve+V3rxxRdrvO+WW24pbb755qV69eqVkpSuuOKKUqlUKu26666lLbbYYrn17brrrqVdd921vH3vvfeWkpT+/Oc/l4YOHVpq2bJlqWHDhqW+ffuWXnnllWXef+6555a+9rWvlaqqqko9evQoPf7448vs85Nq69evX6l9+/Y1xr7zzjulk046qdS2bdtS/fr1SxtvvHHpnHPOKS1ZsqTGuCSlQYMGLVNT+/btS/369Vvu+X7YrFmzSv379y+tu+66pcrKytJWW21Vruuj++vbt++n7m+pyy+/vLTNNtuUqqqqSmuttVZp1113LY0bN67c//DDD5d22mmnUsOGDUtt27YtnXLKKaW///3vpSSle++9tzxu7ty5pUMPPbTUokWLUpIan9OCBQtKv/rVr0pbbLFF+Tjbbbddafjw4aU5c+aUx82bN680aNCg0tprr11q0qRJaf/99y9NmjSplKR09tln16j7ySefLPXu3bvUpEmTUqNGjUq777576ZFHHqkxZun1/Nhjjy1z3kv7pk6dWqP93nvvLfXu3bvUvHnzUoMGDUqdOnUqHXXUUaXHH3+8VCqVSq+//npp0KBBpU033bTUuHHjUvPmzUvdunUrXX/99Z/5MweANVFFqbSKrqIJAADLMWHChGyzzTa56qqrPvG2UQBg1WZNKQAAVlnvvffeMm3nn39+6tSpU2OhdQDgq8eaUgAArLJGjRqVJ554Irvvvnvq1auXO+64I3fccUcGDhy4Qk/UAwBWHW7fAwBglTVu3LgMHz48zz//fObOnZsNNtggRxxxRH7+85+nXj3/vgoAX2VCKQAAAAAKZ00pAAAAAAonlAIAAACgcG7E/wyWLFmS6dOnp2nTpqmoqKjtcgAAAABWWaVSKe+8807atm2bOnU+fj6UUOozmD59uqe7AAAAAHwOr776atZff/2P7RdKfQZNmzZN8sGH2axZs1quBgAAAGDVVV1dnXbt2pXzlI8jlPoMlt6y16xZM6EUAAAAwGfwaUsgWegcAAAAgMLVaig1cuTI7LDDDmnatGlatmyZ/fffP5MmTaox5v3338+gQYOyzjrrpEmTJjnwwAMza9asGmOmTZuWvn37plGjRmnZsmV++tOfZtGiRTXG3Hfffdl2221TVVWVjTbaKGPGjPmyTw8AAACAj1GrodT999+fQYMG5dFHH824ceOycOHC7LXXXpk3b155zEknnZS//e1vueGGG3L//fdn+vTpOeCAA8r9ixcvTt++fbNgwYI88sgj+eMf/5gxY8bk9NNPL4+ZOnVq+vbtm9133z0TJkzIiSeemKOPPjp///vfCz1fAAAAAD5QUSqVSrVdxFKvvfZaWrZsmfvvvz89e/bMnDlzst566+Waa67JQQcdlCR54YUXstlmm2X8+PHZaaedcscdd+Sb3/xmpk+fnlatWiVJLrnkkgwZMiSvvfZaKisrM2TIkNx222157rnnysc65JBD8vbbb+fOO+/81Lqqq6vTvHnzzJkzx5pSAAAAAJ/gs+Yoq9SaUnPmzEmSrL322kmSJ554IgsXLswee+xRHrPppptmgw02yPjx45Mk48ePz1ZbbVUOpJKkd+/eqa6uzsSJE8tjPryPpWOW7gMAAACAYq0yT99bsmRJTjzxxPTo0SNbbrllkmTmzJmprKxMixYtaoxt1apVZs6cWR7z4UBqaf/Svk8aU11dnffeey8NGzas0Td//vzMnz+/vF1dXf3FTxAAAACAslVmptSgQYPy3HPP5dprr63tUjJy5Mg0b968/GrXrl1tlwQAAACwWlklQqnjjz8+t956a+69996sv/765fbWrVtnwYIFefvtt2uMnzVrVlq3bl0e89Gn8S3d/rQxzZo1W2aWVJIMHTo0c+bMKb9effXVL3yOAAAAAPyfWg2lSqVSjj/++Nx0002555570rFjxxr92223XerXr5+777673DZp0qRMmzYt3bt3T5J07949zz77bGbPnl0eM27cuDRr1iybb755ecyH97F0zNJ9fFRVVVWaNWtW4wUAAADAylOrT9877rjjcs011+SWW25J586dy+3Nmzcvz2A69thjc/vtt2fMmDFp1qxZfvSjHyVJHnnkkSTJ4sWL07Vr17Rt2zajRo3KzJkzc8QRR+Too4/OWWedlSSZOnVqttxyywwaNCjf//73c8899+SEE07Ibbfdlt69e39qnZ6+BwAAAPDZfNYcpVZDqYqKiuW2X3HFFTnqqKOSJO+//35OPvnk/PnPf878+fPTu3fv/O53vyvfmpckr7zySo499tjcd999ady4cfr165ezzz479er93zru9913X0466aQ8//zzWX/99XPaaaeVj/FphFIAAAAAn81XIpT6qhBKAQAAAHw2nzVHWSUWOgcAAABgzSKUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAAClevtgugeB1+dlttl7BSvHx239ouAQAAAFhBZkoBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFE0oBAAAAUDihFAAAAACFq9VQ6oEHHsi+++6btm3bpqKiIjfffHON/oqKiuW+zjnnnPKYDh06LNN/9tln19jPM888k1122SUNGjRIu3btMmrUqCJODwAAAICPUauh1Lx587L11lvnoosuWm7/jBkzarwuv/zyVFRU5MADD6wxbsSIETXG/ehHPyr3VVdXZ6+99kr79u3zxBNP5JxzzsmwYcPy+9///ks9NwAAAAA+Xr3aPHifPn3Sp0+fj+1v3bp1je1bbrklu+++ezbccMMa7U2bNl1m7FJXX311FixYkMsvvzyVlZXZYostMmHChIwePToDBw784icBAAAAwOf2lVlTatasWbntttsyYMCAZfrOPvvsrLPOOtlmm21yzjnnZNGiReW+8ePHp2fPnqmsrCy39e7dO5MmTcpbb7213GPNnz8/1dXVNV4AAAAArDy1OlPq8/jjH/+Ypk2b5oADDqjRfsIJJ2TbbbfN2muvnUceeSRDhw7NjBkzMnr06CTJzJkz07FjxxrvadWqVblvrbXWWuZYI0eOzPDhw7+kMwEAAADgKxNKXX755TnssMPSoEGDGu2DBw8u/9ylS5dUVlbmBz/4QUaOHJmqqqoVOtbQoUNr7Le6ujrt2rVbscIBAAAAWMZXIpR68MEHM2nSpFx33XWfOrZbt25ZtGhRXn755XTu3DmtW7fOrFmzaoxZuv1x61BVVVWtcKAFAAAAwKf7Sqwp9Yc//CHbbbddtt56608dO2HChNSpUyctW7ZMknTv3j0PPPBAFi5cWB4zbty4dO7cebm37gEAAADw5avVUGru3LmZMGFCJkyYkCSZOnVqJkyYkGnTppXHVFdX54YbbsjRRx+9zPvHjx+f888/P08//XT+/e9/5+qrr85JJ52Uww8/vBw4HXrooamsrMyAAQMyceLEXHfddbngggtq3J4HAAAAQLFq9fa9xx9/PLvvvnt5e2lQ1K9fv4wZMyZJcu2116ZUKuV73/veMu+vqqrKtddem2HDhmX+/Pnp2LFjTjrppBqBU/PmzTN27NgMGjQo2223XdZdd92cfvrpGThw4Jd7cgAAAAB8rIpSqVSq7SJWddXV1WnevHnmzJmTZs2a1XY5X1iHn91W2yWsFC+f3be2SwAAAAA+4rPmKF+JNaUAAAAAWL0IpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMIJpQAAAAAonFAKAAAAgMLVaij1wAMPZN99903btm1TUVGRm2++uUb/UUcdlYqKihqvvffeu8aYN998M4cddliaNWuWFi1aZMCAAZk7d26NMc8880x22WWXNGjQIO3atcuoUaO+7FMDAAAA4BPUaig1b968bL311rnooos+dszee++dGTNmlF9//vOfa/QfdthhmThxYsaNG5dbb701DzzwQAYOHFjur66uzl577ZX27dvniSeeyDnnnJNhw4bl97///Zd2XgAAAAB8snq1efA+ffqkT58+nzimqqoqrVu3Xm7fv/71r9x555157LHHsv322ydJfvvb32afffbJr3/967Rt2zZXX311FixYkMsvvzyVlZXZYostMmHChIwePbpGeAUAAABAcVb5NaXuu+++tGzZMp07d86xxx6bN954o9w3fvz4tGjRohxIJckee+yROnXq5B//+Ed5TM+ePVNZWVke07t370yaNClvvfVWcScCAAAAQFmtzpT6NHvvvXcOOOCAdOzYMVOmTMmpp56aPn36ZPz48albt25mzpyZli1b1nhPvXr1svbaa2fmzJlJkpkzZ6Zjx441xrRq1arct9Zaay1z3Pnz52f+/Pnl7erq6pV9agAAAABrtFU6lDrkkEPKP2+11Vbp0qVLOnXqlPvuuy+9evX60o47cuTIDB8+/EvbPwAAAMCabpW/fe/DNtxww6y77rqZPHlykqR169aZPXt2jTGLFi3Km2++WV6HqnXr1pk1a1aNMUu3P26tqqFDh2bOnDnl16uvvrqyTwUAAABgjfaVCqX+85//5I033kibNm2SJN27d8/bb7+dJ554ojzmnnvuyZIlS9KtW7fymAceeCALFy4sjxk3blw6d+683Fv3kg8WV2/WrFmNFwAAAAArT62GUnPnzs2ECRMyYcKEJMnUqVMzYcKETJs2LXPnzs1Pf/rTPProo3n55Zdz991351vf+lY22mij9O7dO0my2WabZe+9984xxxyTf/7zn3n44Ydz/PHH55BDDknbtm2TJIceemgqKyszYMCATJw4Mdddd10uuOCCDB48uLZOGwAAAGCNV6uh1OOPP55tttkm22yzTZJk8ODB2WabbXL66aenbt26eeaZZ7Lffvtlk002yYABA7LddtvlwQcfTFVVVXkfV199dTbddNP06tUr++yzT3beeef8/ve/L/c3b948Y8eOzdSpU7Pddtvl5JNPzumnn56BAwcWfr4AAAAAfKCiVCqVaruIVV11dXWaN2+eOXPmrBa38nX42W21XcJK8fLZfWu7BAAAAOAjPmuO8pVaUwoAAACA1YNQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDC1avtAoDVQ4ef3VbbJaw0L5/dt7ZLAAAAWO2ZKQUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4YRSAAAAABROKAUAAABA4Wo1lHrggQey7777pm3btqmoqMjNN99c7lu4cGGGDBmSrbbaKo0bN07btm1z5JFHZvr06TX20aFDh1RUVNR4nX322TXGPPPMM9lll13SoEGDtGvXLqNGjSri9AAAAAD4GLUaSs2bNy9bb711LrroomX63n333Tz55JM57bTT8uSTT+bGG2/MpEmTst9++y0zdsSIEZkxY0b59aMf/ajcV11dnb322ivt27fPE088kXPOOSfDhg3L73//+y/13AAAAAD4ePVq8+B9+vRJnz59ltvXvHnzjBs3rkbbhRdemB133DHTpk3LBhtsUG5v2rRpWrduvdz9XH311VmwYEEuv/zyVFZWZosttsiECRMyevToDBw4cOWdDAAAAACf2VdqTak5c+akoqIiLVq0qNF+9tlnZ5111sk222yTc845J4sWLSr3jR8/Pj179kxlZWW5rXfv3pk0aVLeeuut5R5n/vz5qa6urvECAAAAYOWp1ZlSn8f777+fIUOG5Hvf+16aNWtWbj/hhBOy7bbbZu21184jjzySoUOHZsaMGRk9enSSZObMmenYsWONfbVq1arct9Zaay1zrJEjR2b48OFf4tkAAAAArNm+EqHUwoUL893vfjelUikXX3xxjb7BgweXf+7SpUsqKyvzgx/8ICNHjkxVVdUKHW/o0KE19ltdXZ127dqtWPEAAAAALGOVD6WWBlKvvPJK7rnnnhqzpJanW7duWbRoUV5++eV07tw5rVu3zqxZs2qMWbr9cetQVVVVrXCgBQAAAMCnW6XXlFoaSL300ku56667ss4663zqeyZMmJA6deqkZcuWSZLu3bvngQceyMKFC8tjxo0bl86dOy/31j0AAAAAvny1OlNq7ty5mTx5cnl76tSpmTBhQtZee+20adMmBx10UJ588snceuutWbx4cWbOnJkkWXvttVNZWZnx48fnH//4R3bfffc0bdo048ePz0knnZTDDz+8HDgdeuihGT58eAYMGJAhQ4bkueeeywUXXJDzzjuvVs4ZAAAAgFoOpR5//PHsvvvu5e2l6zj169cvw4YNy1//+tckSdeuXWu87957781uu+2WqqqqXHvttRk2bFjmz5+fjh075qSTTqqxHlTz5s0zduzYDBo0KNttt13WXXfdnH766Rk4cOCXf4IAAAAALFethlK77bZbSqXSx/Z/Ul+SbLvttnn00Uc/9ThdunTJgw8++LnrAwAAAODLsUqvKQUAAADA6kkoBQAAAEDhhFIAAAAAFE4oBQAAAEDhhFIAAAAAFE4oBQAAAEDhhFIAAAAAFE4oBQAAAEDhhFIAAAAAFE4oBQAAAEDh6tV2AQDwSTr87LbaLmGlePnsvrVdAgAArFLMlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAq3QqHUv//975VdBwAAAABrkBUKpTbaaKPsvvvuueqqq/L++++v7JoAAAAAWM2tUCj15JNPpkuXLhk8eHBat26dH/zgB/nnP/+5smsDAAAAYDW1QqFU165dc8EFF2T69Om5/PLLM2PGjOy8887ZcsstM3r06Lz22msru04AAAAAViNfaKHzevXq5YADDsgNN9yQX/3qV5k8eXJ+8pOfpF27djnyyCMzY8aMlVUnAAAAAKuRLxRKPf744znuuOPSpk2bjB49Oj/5yU8yZcqUjBs3LtOnT8+3vvWtlVUnAAAAAKuReivyptGjR+eKK67IpEmTss8+++RPf/pT9tlnn9Sp80HG1bFjx4wZMyYdOnRYmbUCAAAAsJpYoVDq4osvzve///0cddRRadOmzXLHtGzZMn/4wx++UHEAAAAArJ5WKJR66aWXPnVMZWVl+vXrtyK7BwAAAGA1t0JrSl1xxRW54YYblmm/4YYb8sc//vELFwUAAADA6m2FQqmRI0dm3XXXXaa9ZcuWOeuss75wUQAAAACs3lYolJo2bVo6duy4THv79u0zbdq0L1wUAAAAAKu3FQqlWrZsmWeeeWaZ9qeffjrrrLPOFy4KAAAAgNXbCoVS3/ve93LCCSfk3nvvzeLFi7N48eLcc889+fGPf5xDDjlkZdcIAAAAwGpmhZ6+d+aZZ+bll19Or169Uq/eB7tYsmRJjjzySGtKAQAAAPCpViiUqqyszHXXXZczzzwzTz/9dBo2bJitttoq7du3X9n1AQAAALAaWqFQaqlNNtkkm2yyycqqBQAAAIA1xAqFUosXL86YMWNy9913Z/bs2VmyZEmN/nvuuWelFAcAAADA6mmFQqkf//jHGTNmTPr27Zstt9wyFRUVK7suAAAAAFZjKxRKXXvttbn++uuzzz77rOx6AAAAAFgD1FmRN1VWVmajjTZa2bUAAAAAsIZYoVDq5JNPzgUXXJBSqbSy6wEAAABgDbBCt+899NBDuffee3PHHXdkiy22SP369Wv033jjjSulOAAAAABWTysUSrVo0SLf/va3V3YtAAAAAKwhViiUuuKKK1Z2HQAAAACsQVZoTakkWbRoUe66665ceumleeedd5Ik06dPz9y5c1dacQAAAACsnlZoptQrr7ySvffeO9OmTcv8+fOz5557pmnTpvnVr36V+fPn55JLLlnZdQIAAACwGlmhmVI//vGPs/322+ett95Kw4YNy+3f/va3c/fdd6+04gAAAABYPa3QTKkHH3wwjzzySCorK2u0d+jQIf/9739XSmEAAAAArL5WaKbUkiVLsnjx4mXa//Of/6Rp06ZfuCgAAAAAVm8rFErttddeOf/888vbFRUVmTt3bs4444zss88+K6s2AAAAAFZTK3T73rnnnpvevXtn8803z/vvv59DDz00L730UtZdd938+c9/Xtk1AgAAALCaWaGZUuuvv36efvrpnHrqqTnppJOyzTbb5Oyzz85TTz2Vli1bfub9PPDAA9l3333Ttm3bVFRU5Oabb67RXyqVcvrpp6dNmzZp2LBh9thjj7z00ks1xrz55ps57LDD0qxZs7Ro0SIDBgzI3Llza4x55plnsssuu6RBgwZp165dRo0atSKnDQAAAMBKskIzpZKkXr16Ofzww7/QwefNm5ett9463//+93PAAQcs0z9q1Kj85je/yR//+Md07Ngxp512Wnr37p3nn38+DRo0SJIcdthhmTFjRsaNG5eFCxemf//+GThwYK655pokSXV1dfbaa6/sscceueSSS/Lss8/m+9//flq0aJGBAwd+ofoBAAAAWDErFEr96U9/+sT+I4888jPtp0+fPunTp89y+0qlUs4///z84he/yLe+9a3ycVu1apWbb745hxxySP71r3/lzjvvzGOPPZbtt98+SfLb3/42++yzT37961+nbdu2ufrqq7NgwYJcfvnlqayszBZbbJEJEyZk9OjRQikAAACAWrJCodSPf/zjGtsLFy7Mu+++m8rKyjRq1Ogzh1KfZOrUqZk5c2b22GOPclvz5s3TrVu3jB8/PoccckjGjx+fFi1alAOpJNljjz1Sp06d/OMf/8i3v/3tjB8/Pj179kxlZWV5TO/evfOrX/0qb731VtZaa60vXCsAAAAAn88KrSn11ltv1XjNnTs3kyZNys4777zSFjqfOXNmkqRVq1Y12lu1alXumzlz5jJrWNWrVy9rr712jTHL28eHj/FR8+fPT3V1dY0XAAAAACvPCoVSy7Pxxhvn7LPPXmYW1VfRyJEj07x58/KrXbt2tV0SAAAAwGplpYVSyQezlKZPn75S9tW6deskyaxZs2q0z5o1q9zXunXrzJ49u0b/okWL8uabb9YYs7x9fPgYHzV06NDMmTOn/Hr11Ve/+AkBAAAAULZCa0r99a9/rbFdKpUyY8aMXHjhhenRo8dKKaxjx45p3bp17r777nTt2jXJB0/S+8c//pFjjz02SdK9e/e8/fbbeeKJJ7LddtslSe65554sWbIk3bp1K4/5+c9/noULF6Z+/fpJknHjxqVz584fu55UVVVVqqqqVsp5AAAAALCsFQql9t9//xrbFRUVWW+99fKNb3wj55577mfez9y5czN58uTy9tSpUzNhwoSsvfba2WCDDXLiiSfml7/8ZTbeeON07Ngxp512Wtq2bVs+/mabbZa99947xxxzTC655JIsXLgwxx9/fA455JC0bds2SXLooYdm+PDhGTBgQIYMGZLnnnsuF1xwQc4777wVOXUAAAAAVoIVCqWWLFmyUg7++OOPZ/fddy9vDx48OEnSr1+/jBkzJqecckrmzZuXgQMH5u23387OO++cO++8Mw0aNCi/5+qrr87xxx+fXr16pU6dOjnwwAPzm9/8ptzfvHnzjB07NoMGDcp2222XddddN6effnoGDhy4Us4BAAAAgM9vhUKplWW33XZLqVT62P6KioqMGDEiI0aM+Ngxa6+9dq655ppPPE6XLl3y4IMPrnCdAAAAAKxcKxRKLZ3R9FmMHj16RQ4BAAAAwGpshUKpp556Kk899VQWLlyYzp07J0lefPHF1K1bN9tuu215XEVFxcqpEgAAAIDVygqFUvvuu2+aNm2aP/7xj+Un2L311lvp379/dtlll5x88skrtUgAAAAAVi91VuRN5557bkaOHFkOpJJkrbXWyi9/+cvP9fQ9AAAAANZMKxRKVVdX57XXXlum/bXXXss777zzhYsCAAAAYPW2QqHUt7/97fTv3z833nhj/vOf/+Q///lP/vd//zcDBgzIAQccsLJrBAAAAGA1s0JrSl1yySX5yU9+kkMPPTQLFy78YEf16mXAgAE555xzVmqBAAAAAKx+ViiUatSoUX73u9/lnHPOyZQpU5IknTp1SuPGjVdqcQAAAACsnlbo9r2lZsyYkRkzZmTjjTdO48aNUyqVVlZdAAAAAKzGViiUeuONN9KrV69ssskm2WeffTJjxowkyYABA3LyySev1AIBAAAAWP2sUCh10kknpX79+pk2bVoaNWpUbj/44INz5513rrTiAAAAAFg9rdCaUmPHjs3f//73rL/++jXaN95447zyyisrpTAAAAAAVl8rNFNq3rx5NWZILfXmm2+mqqrqCxcFAAAAwOpthUKpXXbZJX/605/K2xUVFVmyZElGjRqV3XfffaUVBwAAAMDqaYVu3xs1alR69eqVxx9/PAsWLMgpp5ySiRMn5s0338zDDz+8smsEAAAAYDWzQjOlttxyy7z44ovZeeed861vfSvz5s3LAQcckKeeeiqdOnVa2TUCAAAAsJr53DOlFi5cmL333juXXHJJfv7zn38ZNQEAAACwmvvcM6Xq16+fZ5555suoBQAAAIA1xArdvnf44YfnD3/4w8quBQAAAIA1xAotdL5o0aJcfvnlueuuu7LddtulcePGNfpHjx69UooDAAAAYPX0uUKpf//73+nQoUOee+65bLvttkmSF198scaYioqKlVcdAAAAAKulzxVKbbzxxpkxY0buvffeJMnBBx+c3/zmN2nVqtWXUhwAAAAAq6fPtaZUqVSqsX3HHXdk3rx5K7UgAAAAAFZ/K7TQ+VIfDakAAAAA4LP4XKFURUXFMmtGWUMKAAAAgM/rc60pVSqVctRRR6WqqipJ8v777+eHP/zhMk/fu/HGG1dehQAAAACsdj5XKNWvX78a24cffvhKLQYAAACANcPnCqWuuOKKL6sOAAAAANYgX2ihcwAAAABYEUIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAonlAIAAACgcEIpAAAAAAq3yodSHTp0SEVFxTKvQYMGJUl22223Zfp++MMf1tjHtGnT0rdv3zRq1CgtW7bMT3/60yxatKg2TgcAAACAJPVqu4BP89hjj2Xx4sXl7eeeey577rlnvvOd75TbjjnmmIwYMaK83ahRo/LPixcvTt++fdO6des88sgjmTFjRo488sjUr18/Z511VjEnAQAAAEANq3wotd5669XYPvvss9OpU6fsuuuu5bZGjRqldevWy33/2LFj8/zzz+euu+5Kq1at0rVr15x55pkZMmRIhg0blsrKyi+1fgAAAACWtcrfvvdhCxYsyFVXXZXvf//7qaioKLdfffXVWXfddbPllltm6NCheffdd8t948ePz1ZbbZVWrVqV23r37p3q6upMnDix0PoBAAAA+MAqP1Pqw26++ea8/fbbOeqoo8pthx56aNq3b5+2bdvmmWeeyZAhQzJp0qTceOONSZKZM2fWCKSSlLdnzpy53OPMnz8/8+fPL29XV1ev5DMBAAAAWLN9pUKpP/zhD+nTp0/atm1bbhs4cGD556222ipt2rRJr169MmXKlHTq1GmFjjNy5MgMHz78C9cLAAAAwPJ9ZW7fe+WVV3LXXXfl6KOP/sRx3bp1S5JMnjw5SdK6devMmjWrxpil2x+3DtXQoUMzZ86c8uvVV1/9ouUDAAAA8CFfmVDqiiuuSMuWLdO3b99PHDdhwoQkSZs2bZIk3bt3z7PPPpvZs2eXx4wbNy7NmjXL5ptvvtx9VFVVpVmzZjVeAAAAAKw8X4nb95YsWZIrrrgi/fr1S716/1fylClTcs0112SfffbJOuusk2eeeSYnnXRSevbsmS5duiRJ9tprr2y++eY54ogjMmrUqMycOTO/+MUvMmjQoFRVVdXWKQEAAACs0b4SodRdd92VadOm5fvf/36N9srKytx11105//zzM2/evLRr1y4HHnhgfvGLX5TH1K1bN7feemuOPfbYdO/ePY0bN06/fv0yYsSIok8DAAAAgP/fVyKU2muvvVIqlZZpb9euXe6///5PfX/79u1z++23fxmlAQAAALACvjJrSgEAAACw+hBKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFA4oRQAAAAAhRNKAQAAAFC4VTqUGjZsWCoqKmq8Nt1003L/+++/n0GDBmWdddZJkyZNcuCBB2bWrFk19jFt2rT07ds3jRo1SsuWLfPTn/40ixYtKvpUAAAAAPiQerVdwKfZYostctddd5W369X7v5JPOumk3HbbbbnhhhvSvHnzHH/88TnggAPy8MMPJ0kWL16cvn37pnXr1nnkkUcyY8aMHHnkkalfv37OOuusws8FAAAAgA+s8qFUvXr10rp162Xa58yZkz/84Q+55ppr8o1vfCNJcsUVV2SzzTbLo48+mp122iljx47N888/n7vuuiutWrVK165dc+aZZ2bIkCEZNmxYKisriz4dAAAAALKK376XJC+99FLatm2bDTfcMIcddlimTZuWJHniiSeycOHC7LHHHuWxm266aTbYYIOMHz8+STJ+/PhstdVWadWqVXlM7969U11dnYkTJ37sMefPn5/q6uoaLwAAAABWnlU6lOrWrVvGjBmTO++8MxdffHGmTp2aXXbZJe+8805mzpyZysrKtGjRosZ7WrVqlZkzZyZJZs6cWSOQWtq/tO/jjBw5Ms2bNy+/2rVrt3JPDAAAAGANt0rfvtenT5/yz126dEm3bt3Svn37XH/99WnYsOGXdtyhQ4dm8ODB5e3q6mrBFAAAAMBKtErPlPqoFi1aZJNNNsnkyZPTunXrLFiwIG+//XaNMbNmzSqvQdW6detlnsa3dHt561QtVVVVlWbNmtV4AQAAALDyfKVCqblz52bKlClp06ZNtttuu9SvXz933313uX/SpEmZNm1aunfvniTp3r17nn322cyePbs8Zty4cWnWrFk233zzwusHAAAA4AOr9O17P/nJT7Lvvvumffv2mT59es4444zUrVs33/ve99K8efMMGDAggwcPztprr51mzZrlRz/6Ubp3756ddtopSbLXXntl8803zxFHHJFRo0Zl5syZ+cUvfpFBgwalqqqqls8OAAAAYM21SodS//nPf/K9730vb7zxRtZbb73svPPOefTRR7PeeuslSc4777zUqVMnBx54YObPn5/evXvnd7/7Xfn9devWza233ppjjz023bt3T+PGjdOvX7+MGDGitk4JAAAAgKziodS11177if0NGjTIRRddlIsuuuhjx7Rv3z633377yi4NAAAAgC/gK7WmFAAAAACrB6EUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQOKEUAAAAAIUTSgEAAABQuFU6lBo5cmR22GGHNG3aNC1btsz++++fSZMm1Riz2267paKiosbrhz/8YY0x06ZNS9++fdOoUaO0bNkyP/3pT7No0aIiTwUAAACAD6lX2wV8kvvvvz+DBg3KDjvskEWLFuXUU0/NXnvtleeffz6NGzcujzvmmGMyYsSI8najRo3KPy9evDh9+/ZN69at88gjj2TGjBk58sgjU79+/Zx11lmFng8AAAAAH1ilQ6k777yzxvaYMWPSsmXLPPHEE+nZs2e5vVGjRmnduvVy9zF27Ng8//zzueuuu9KqVat07do1Z555ZoYMGZJhw4alsrLySz0HAAAAAJa1St++91Fz5sxJkqy99to12q+++uqsu+662XLLLTN06NC8++675b7x48dnq622SqtWrcptvXv3TnV1dSZOnFhM4QAAAADUsErPlPqwJUuW5MQTT0yPHj2y5ZZbltsPPfTQtG/fPm3bts0zzzyTIUOGZNKkSbnxxhuTJDNnzqwRSCUpb8+cOXO5x5o/f37mz59f3q6url7ZpwMAAACwRvvKhFKDBg3Kc889l4ceeqhG+8CBA8s/b7XVVmnTpk169eqVKVOmpFOnTit0rJEjR2b48OFfqF4AAAAAPt5X4va9448/PrfeemvuvfferL/++p84tlu3bkmSyZMnJ0lat26dWbNm1RizdPvj1qEaOnRo5syZU369+uqrX/QUAAAAAPiQVTqUKpVKOf7443PTTTflnnvuSceOHT/1PRMmTEiStGnTJknSvXv3PPvss5k9e3Z5zLhx49KsWbNsvvnmy91HVVVVmjVrVuMFAAAAwMqzSt++N2jQoFxzzTW55ZZb0rRp0/IaUM2bN0/Dhg0zZcqUXHPNNdlnn32yzjrr5JlnnslJJ52Unj17pkuXLkmSvfbaK5tvvnmOOOKIjBo1KjNnzswvfvGLDBo0KFVVVbV5egAAAABrrFV6ptTFF1+cOXPmZLfddkubNm3Kr+uuuy5JUllZmbvuuit77bVXNt1005x88sk58MAD87e//a28j7p16+bWW29N3bp107179xx++OE58sgjM2LEiNo6LQAAAIA13io9U6pUKn1if7t27XL//fd/6n7at2+f22+/fWWVBQAAAMAXtErPlAIAAABg9SSUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAACieUAgAAAKBwQikAAAAAClevtgsAAABWPR1+dlttl7BSvHx239ouAYCPYaYUAAAAAIUTSgEAAABQOKEUAAAAAIWzphQAwFeYdX8AgK8qM6UAAAAAKJxQCgAAAIDCCaUAAAAAKJxQCgAAAIDCWegcAACAwnhAA7CUmVIAAAAAFE4oBQAAAEDhhFIAAAAAFE4oBQAAAEDhhFIAAAAAFG6NCqUuuuiidOjQIQ0aNEi3bt3yz3/+s7ZLAgAAAFgjrTGh1HXXXZfBgwfnjDPOyJNPPpmtt946vXv3zuzZs2u7NAAAAIA1Tr3aLqAoo0ePzjHHHJP+/fsnSS655JLcdtttufzyy/Ozn/2slqsDAAAAVlSHn91W2yWsNC+f3be2SyjMGhFKLViwIE888USGDh1abqtTp0722GOPjB8/fpnx8+fPz/z588vbc+bMSZJUV1d/+cUWYMn8d2u7hJVidfnvsbpYXa6rxLW1qlldri3XFV8W/4/wZXFt8WVxbfFlWF2uq2T1uLaWnkOpVPrEcWtEKPX6669n8eLFadWqVY32Vq1a5YUXXlhm/MiRIzN8+PBl2tu1a/el1cjn1/z82q6A1ZVriy+D6wo+mf9H+LK4tviyuLb4sqxO19Y777yT5s2bf2z/GhFKfV5Dhw7N4MGDy9tLlizJm2++mXXWWScVFRW1WNkXV11dnXbt2uXVV19Ns2bNarscKHNtsipzfbKqcm2yKnN9sqpybbIqW12uz1KplHfeeSdt27b9xHFrRCi17rrrpm7dupk1a1aN9lmzZqV169bLjK+qqkpVVVWNthYtWnyZJRauWbNmX+kLnNWXa5NVmeuTVZVrk1WZ65NVlWuTVdnqcH1+0gyppdaIp+9VVlZmu+22y913311uW7JkSe6+++507969FisDAAAAWDOtETOlkmTw4MHp169ftt9+++y44445//zzM2/evPLT+AAAAAAozhoTSh188MF57bXXcvrpp2fmzJnp2rVr7rzzzmUWP1/dVVVV5Ywzzljm9kSoba5NVmWuT1ZVrk1WZa5PVlWuTVZla9r1WVH6tOfzAQAAAMBKtkasKQUAAADAqkUoBQAAAEDhhFIAAAAAFE4oBQAAAEDhhFIAAAAAFE4oBRTCgz4BYPXx7LPPprq6urbLAOArrl5tF8DKMW7cuPztb3/LhhtumK9//evZcccda7skSJLccccdufzyy9OqVav06NEjBx54YCorK2u7LEiSPPDAA3nvvffSq1ev1Kvnj0RWHbfeemv+8Y9/ZJNNNknPnj3Tvn372i4JkiR//vOfM2LEiNSvXz9z587NiSeemBNOOKG2y4IkyY033pi//e1vadeuXfbee+98/etfr+2SoIZ///vf2WCDDfze+SFmSn3FzZgxI/vuu28OP/zwvPnmm7n88suz11575Z///Gdtl8Ya7r///W/69u2bfv365Wtf+1pmzpyZY445JjfffHNtlwZ5/fXX069fv+y222455ZRT8p///Ke2S4IkyZQpU7Lrrrvm2GOPzXPPPZdTTz01ffr0ybx582q7NNZws2fPzmGHHZaf/OQn+fGPf5yLL744RxxxRH7605/mjTfeqO3yWMO99NJL6dmzZ370ox+lsrIyY8eOTd++ffP888/XdmmQJLnsssuyxRZb5Dvf+U522WWX3HHHHUmSJUuW1HJltU889xX27rvvZujQoWncuHEeffTRdOzYMUnSrVu3XHrppdlxxx2zZMmS1Kkje6RY7777bs4666w0b948Tz75ZNZff/0kSZcuXfLII4/ku9/9bi1XyJps0aJFueGGGzJr1qxce+216d+/f6699toMHjzYLD5q1dSpUzNo0KBsvPHGuf7667POOutk9uzZ2XrrrXPZZZflxBNPTKlUSkVFRW2XyhpoypQpqaioyNixY7PFFlskSVq1apVrrrkmb775ZtZZZx3XJ7ViypQp+clPfpLOnTvnpptuyjrrrJMkadasWR5++OFsvvnmtVwha7IpU6Zk0KBBeeGFFzJs2LA0a9Ysf/nLX3L00Ufnv//9r7+rRyj1ldaoUaNUVVXlkEMOSceOHbNo0aLUq1cv++yzTzl5dZFTGxo1apTvfOc7WX/99cuBVJJsvPHG2WeffTJ//vxUVVXVYoWsyerVq5dtt90266+/fvbdd9+88MILGT16dPbee+907dq1tstjDdaiRYuUSqX84Ac/SKtWrVIqldKmTZvsuuuu5X/t9xd+akv37t3TrFmzGn/Bv+iii9KqVatMmDAhTZo0SZs2bWqxQtZUHTp0yMEHH5yePXuWA6k333wzPXr0yC677FLL1bGmmzBhQho0aJAHH3ww7dq1S/LBWrsvvfRS3n777fKf/Wvyn+8VJasPf6UtXLgw9evXT5LyrKjDDjssjRs3zu9///s1/gKn9nz42rvvvvty9NFHZ8aMGenUqVPWXnvtHH/88TnooINquUrWVB/9bvza176Wb37zm/n1r3+dpk2b1mJlrEmWt6bZ0l9Qk/+7Trt165ZDDjkkJ510Ui1Wy5rk09Yq/fe//50DDjggs2bNSs+ePfPCCy9k/vz5ueiii9KrV69aqpo1xfK+O5f+43zywZp8P/zhDzN37txssMEG2XLLLXPyySdnu+22q82yWcN8+HfNZ555Jl26dCn3HXHEEVm4cGGOO+647LDDDmnYsGFtlblKMI3mK25pIJX836yoV155JT169KitkiDJ//1r/sKFCzN27Njst99+mTBhQq688sq0adMm559/vvv8qTVLr88FCxYkSS644IJcfvnlefTRR2uzLNYQH17TbMiQIXn11VfLfUv/xTT54Dp944038sYbb3iACYX4uLVKH3vssRrjFi5cmNNPPz0vvfRS/vznP+fpp59OqVTKTTfdlMQTd/lyfPS788PrQS4NpKZPn57rrrsu/fv3z4MPPpjzzz8/zz33XEaPHm3tM750t956a0477bRceeWVmTZtWrl9aSD1z3/+Mx07dswDDzyQOXPmZMCAAendu3f++9//1lbJqwSh1Grm3//+dyZPnpwtt9wyyQe/0C5cuLCWq2JNVr9+/YwYMSKjR49Op06dsvXWW+eII47I5MmTs2jRotoujzXc0jWkDjrooOywww4ZNWpUZs+enSSZOXNmbZbGauqja5q99NJLue6668oBaVLzFr2nnnoqCxcuzEYbbVRuW7x4caE1s2b46FqlV111VZ555pl07tw5l1xySZKU/9zu3LlzDjjggDRp0qS8SO+WW26Zp556KonbTFn5lvfdee2115a/O5cGoW3atMlll12WM888M5tvvnm+8Y1vpH///nn00UctHcGX5uMeUvLuu+/WGDdv3rycddZZ+de//pWbb745jz/+eB577LH89a9/TbLmBvrWlFpNLJ0e+NBDD6VJkybl6anDhw/PzJkzM3z48LRs2bKWq2RNtfRfr5b+kjp9+vSstdZaWXfddWuzLEjyf1P+L7vssmy99da59tprM2XKlIwfPz6XXnppttlmm9oukdXIZ13TbPHixalbt27uueeebLLJJmnVqlUWLFiQU089Na1bt86Pf/zjGrOl4Yv6LGuVLu8R5vXq1cukSZMyc+bMHH/88UWXzRri0747l/6OWVFRkQYNGpR/LpVKmT59etq2bSss5UuxvIeUzJo1K127ds3vf//7nHjiieXv0913373Ge6uqqrLpppvm8ccfT7LmBvpmSq0mll7A//znP3PggQdm3Lhx6dixY373u9/l29/+tkCKWrdw4cJUVFTkf//3f3PBBRfkiCOOSNu2bWu7LCj/JWuLLbbItttumxNPPDG33HJLzjzzTIEUX4odd9wx++67b5Lk9NNPT/369XPxxRfnnXfeKY+pW7dukuTJJ5/MwQcfnL/+9a/p2LFjrrnmmuyyyy4CKb4UF154Yfbee+8k/7csxIsvvli+9eTD/4r/+uuv56WXXsqVV16ZPn36pHXr1taT4kv1Wb47l1q8eHHq1KmTK6+8MrfffnuOOeaYNG7cuOiSWQN89CEldevWTdu2bWs8pGR5gX6SPPLII6moqEj//v2LLHnVU2K18d5775U22mijUkVFRamqqqp09tln13ZJUCqVSqU333yzdNJJJ5X23HPPUrNmzUq//e1va7skqGHy5MmlLbfcstSoUaPS//t//6+2y2ENMX/+/FKpVCpdf/31pXr16pXGjh1b7luyZElp0qRJpaqqqlL9+vVLjRo1Kl144YW1VSprsB49epTGjBlTKpU+uC5LpVJpwYIFpRtvvLG0/fbblzp06FC6+OKLa7NE1jCf9N1ZKpVKTzzxRGnw4MGlXXbZpbTOOuuU/vjHP9ZGmaym7r///tKdd95ZWrhwYbntrbfeKv+89Htyxx13LI0ePXqZ97/22mulV155pXTxxReXNthgg9IxxxxTqq6u/tLrXpW5fW810qBBg3To0CF77rlnRo8eXZ66CrVtrbXWygYbbJDmzZvnr3/9q2uTVU7dunVz4IEHZsiQIWv8E1AoztI1zb7zne/kvPPOy6hRo7L11lunZcuWeeONN9KiRYu0aNEi/fv3z8iRI2u5WtZEy1urdNGiRalfv3723HPPVFVVZZ999qnlKlnTfNJ352uvvZbOnTtn7ty52WOPPfLAAw/UcrWsLl5//fWcfPLJufLKK9OlS5fcdNNN6dixY5L/my1VUVHxiQ8pefPNN3PzzTfnoosuyttvv52zzjorhx12WG2cziqlolRaQ1fTWk0tXYMCVjWlDz0WFYAPLF1nYuLEidl6660zevToTJkyJQ8//HD+8Ic/ZKONNnLLCYVb+mf2n/70p4wYMSKTJ09O8sFapTNmzMjw4cPTqlWrWq6SNdnHfXc+9NBDueqqq7Lxxht/7C1T8HktWrQol112WW655ZZ8//vfT//+/XPaaadl8ODB5ZD0w+66664MGDAg//znP8vflUu/VydPnpznnnsu+++/f8FnseoSSgEArAJ23HHHPP7442nXrl1+//vfp3fv3rVdEmu4448/Po0bN84ee+yRgQMH5t13382VV16Zvfbaq7ZLg7Kl350bbLBBLr30Ut+dfCn+8Y9/ZPbs2dl3330zYsSIXHjhhRk7duxyH1Jy6qmn5rHHHsu4cePKDylp1apVTjzxRGtCLodQCgCgFk2ZMiX7779//v3vf+c3v/lNBgwYUNslQd5///1stdVWmTJlSiorKzN8+PAMGTKktsuCMt+dFOmjd3187Wtfyze/+c38+te/TtOmTWuM3XvvvXPQQQelZcuWOfbYY1MqlXLTTTelW7duRZf9lSCUAgCoRS+//HLGjBljTTNWOXvuuWc23nhja5WySvLdSW1YsGBBKisrc8MNN+TQQw/N7bffnj333DPJB8HVSy+9lC5dumTJkiWpX79+Ro0alUGDBtVy1as2oRQAALAMa5UCfLyvf/3rady4ca6++uq0bNkyr7/+epYsWZIuXbp4SMnnIJQCAAAA+Aw8pGTlEkoBAAAAfE4eUvLF1antAgAAAAC+KqZMmZKtttoqEydOzGWXXZZXXnlFILWChFIAAAAAn1HdunVz4IEH5vXXX/fkxy/I7XsAAAAAFM5MKQAAAAAKJ5QCAAAAoHBCKQAAAAAKJ5QCAAAAoHBCKQAAAAAKJ5QCAAAAoHBCKQAAAAAKJ5QCACBjxoxJixYtarsMAGANIpQCACjYsGHD0rVr19ouo4aDDz44L774Ym2XAQCsQerVdgEAANSuhQsXpmHDhmnYsGFtlwIArEHMlAIAWAFLlizJqFGjstFGG6WqqiobbLBB/ud//idJMmTIkGyyySZp1KhRNtxww5x22mlZuHBhkg9ukxs+fHiefvrpVFRUpKKiImPGjEmSvP322zn66KOz3nrrpVmzZvnGN76Rp59+usZxf/nLX6Zly5Zp2rRpjj766PzsZz+rMetqyZIlGTFiRNZff/1UVVWla9euufPOO8v9L7/8cioqKnLddddl1113TYMGDXL11Vcv9/a9W265Jdtuu20aNGiQDTfcMMOHD8+iRYuSJKVSKcOGDcsGG2yQqqqqtG3bNieccMJK/pQBgNWZmVIAACtg6NChueyyy3Leeedl5513zowZM/LCCy8kSZo2bZoxY8akbdu2efbZZ3PMMcekadOmOeWUU3LwwQfnueeey5133pm77rorSdK8efMkyXe+8500bNgwd9xxR5o3b55LL700vXr1yosvvpi11147V199df7nf/4nv/vd79KjR49ce+21Offcc9OxY8dyXRdccEHOPffcXHrppdlmm21y+eWXZ7/99svEiROz8cYbl8f97Gc/y7nnnpttttkmDRo0yN///vca5/fggw/myCOPzG9+85vssssumTJlSgYOHJgkOeOMM/K///u/Oe+883Lttddmiy22yMyZM5cJ0AAAPklFqVQq1XYRAABfJe+8807WW2+9XHjhhTn66KM/dfyvf/3rXHvttXn88ceTfLCm1M0335wJEyaUxzz00EPp27dvZs+enaqqqnL7RhttlFNOOSUDBw7MTjvtlO233z4XXnhhuX/nnXfO3Llzy/v62te+lkGDBuXUU08tj9lxxx2zww475KKLLsrLL7+cjh075vzzz8+Pf/zj8pgxY8bkxBNPzNtvv50k2WOPPdKrV68MHTq0POaqq67KKaeckunTp2f06NG59NJL89xzz6V+/fqf6/MDAEjcvgcA8Ln961//yvz589OrV6/l9l933XXp0aNHWrdunSZNmuQXv/hFpk2b9on7fPrppzN37tyss846adKkSfk1derUTJkyJUkyadKk7LjjjjXe9+Ht6urqTJ8+PT169KgxpkePHvnXv/5Vo2377bf/1HpGjBhRo5ZjjjkmM2bMyLvvvpvvfOc7ee+997LhhhvmmGOOyU033VS+tQ8A4LNw+x4AwOf0SQuCjx8/PocddliGDx+e3r17p3nz5uXb7D7J3Llz06ZNm9x3333L9H10raeVoXHjxp9az/Dhw3PAAQcs09egQYO0a9cukyZNyl133ZVx48bluOOOyznnnJP777/fzCkA4DMRSgEAfE4bb7xxGjZsmLvvvnuZ2/ceeeSRtG/fPj//+c/Lba+88kqNMZWVlVm8eHGNtm233TYzZ85MvXr10qFDh+Uet3Pnznnsscdy5JFHltsee+yx8s/NmjVL27Zt8/DDD2fXXXcttz/88MPLzLD6NNtuu20mTZqUjTba6GPHNGzYMPvuu2/23XffDBo0KJtuummeffbZbLvttp/rWADAmkkoBQDwOTVo0CBDhgzJKaecksrKyvTo0SOvvfZaeTHxadOm5dprr80OO+yQ2267LTfddFON93fo0CFTp07NhAkTsv7666dp06bZY4890r179+y///4ZNWpUNtlkk0yfPj233XZbvv3tb2f77bfPj370oxxzzDHZfvvt8/Wvfz3XXXddnnnmmWy44Yblff/0pz/NGWeckU6dOqVr16654oorMmHChFx99dWf6xxPP/30fPOb38wGG2yQgw46KHXq1MnTTz+d5557Lr/85S8zZsyYLF68ON26dUujRo1y1VVXpWHDhmnfvv1K+YwBgNWfNaUAAFbAaaedlpNPPjmnn356Nttssxx88MGZPXt29ttvv5x00kk5/vjj07Vr1zzyyCM57bTTarz3wAMPzN57753dd9896623Xv785z+noqIit99+e3r27Jn+/ftnk002ySGHHJJXXnklrVq1SpIcdthhGTp0aH7yk59k2223zdSpU3PUUUelQYMG5X2fcMIJGTx4cE4++eRstdVWufPOO/PXv/61xpP3PovevXvn1ltvzdixY7PDDjtkp512ynnnnVcOnVq0aJHLLrssPXr0SJcuXXLXXXflb3/7W9ZZZ50v+MkCAGsKT98DAPgK23PPPdO6detceeWVtV0KAMDn4vY9AICviHfffTeXXHJJevfunbp16+bPf/5zeaFxAICvGjOlAAC+It57773su+++eeqpp/L++++nc+fO+cUvfrHcJ+QBAKzqhFIAAAAAFM5C5wAAAAAUTigFAAAAQOGEUgAAAAAUTigFAAAAQOGEUgAAAAAUTigFAAAAQOGEUgAAAAAUTigFAAAAQOGEUgAAAAAU7v8DZ2P9nfMJzloAAAAASUVORK5CYII=\n"
          },
          "metadata": {}
        }
      ],
      "execution_count": 121
    },
    {
      "cell_type": "code",
      "source": [
        "sample_contri.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 551
        },
        "id": "3f-zb3AmZlLO",
        "outputId": "26416385-8510-409c-d699-1bebceda7b0a"
      },
      "execution_count": 127,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       idx    Sites           Genus  abund_raw  rel_abund_raw  \\\n",
              "0   502012  site_27          Afipia   267830.0      28.021013   \n",
              "1   562124  site_30   Anoxybacillus    32682.0       4.590262   \n",
              "2   472739  site_25    Sphingomonas     1121.0       0.168339   \n",
              "3   133617   site_7  Staphylococcus     7736.0       0.878632   \n",
              "4  1136015  site_63  Bradyrhizobium     4161.0       0.613867   \n",
              "\n",
              "   genome_EC_count  abund_contri  rel_abund_contri  norm_abund_contri  \\\n",
              "0                2      535660.0         56.042026           0.353589   \n",
              "1                2       65364.0          9.180525           0.054352   \n",
              "2                1        1121.0          0.168339           0.001973   \n",
              "3                1        7736.0          0.878632           0.940890   \n",
              "4                8       33288.0          4.910938           0.011502   \n",
              "\n",
              "                                        protein_name  ...  \\\n",
              "0                             prephenate dehydratase  ...   \n",
              "1  corrinoid adenosyltransferase alamin adenosylt...  ...   \n",
              "2                    phosphoenolpyruvate carboxylase  ...   \n",
              "3  cystathionine gamma-lyase and homocysteine for...  ...   \n",
              "4                acetolactate-synthase small subunit  ...   \n",
              "\n",
              "                                            pathways  \\\n",
              "0  Phenylalanine, tyrosine and tryptophan biosynt...   \n",
              "1  Arginine biosynthesis; Arginine and proline me...   \n",
              "2  Glycolysis / Gluconeogenesis; Citrate cycle (T...   \n",
              "3  Diterpenoid biosynthesis; Biosynthesis of seco...   \n",
              "4  Valine, leucine and isoleucine biosynthesis; B...   \n",
              "\n",
              "                                           hierarchy  \\\n",
              "0  Amino acid metabolism > Phenylalanine, tyrosin...   \n",
              "1  Cardiovascular disease > Viral myocarditis; Ne...   \n",
              "2  Energy metabolism > Carbon fixation by Calvin ...   \n",
              "3  Metabolism of terpenoids and polyketides > Dit...   \n",
              "4  Metabolism of cofactors and vitamins > Pantoth...   \n",
              "\n",
              "                                     metals_involved  \\\n",
              "0  potassium; phosphate; iron; sulfur; hydrogen; ...   \n",
              "1  Mo; iron; zinc; hydrogen; selenium; nitrate; n...   \n",
              "2  Mo; iron; sulfur; hydrogen; sodium; selenium; ...   \n",
              "3                                           selenium   \n",
              "4                Al3+; Mo; iron; selenium; phosphate   \n",
              "\n",
              "                  metals_consolidated  \\\n",
              "0  K; phosphate; H; Mo; Mn; Se; Fe; S   \n",
              "1        H; Zn; Mo; NO3-; Se; Fe; NO2   \n",
              "2     H; phosphate; Mo; Se; Fe; S; Na   \n",
              "3                                  Se   \n",
              "4         phosphate; Mo; Se; Fe; Al3+   \n",
              "\n",
              "                                corrosion_mechanisms  \\\n",
              "0  h2_consumption; acid_production; sulfur_metabo...   \n",
              "1                         h2_consumption; direct_eet   \n",
              "2                    h2_consumption; acid_production   \n",
              "3                                               None   \n",
              "4                                    acid_production   \n",
              "\n",
              "  corrosion_relevance_score corrosion_relevance has_metal Category  \\\n",
              "0                      20.0                high      True        2   \n",
              "1                      14.5                high      True        2   \n",
              "2                      15.5                high      True        2   \n",
              "3                       1.5                 low      True        2   \n",
              "4                      10.5                high      True        2   \n",
              "\n",
              "   was_uncharacterized  \n",
              "0                False  \n",
              "1                False  \n",
              "2                False  \n",
              "3                False  \n",
              "4                False  \n",
              "\n",
              "[5 rows x 24 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e29d1bed-cd29-42bc-a4af-77a37e2e7e32\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>idx</th>\n",
              "      <th>Sites</th>\n",
              "      <th>Genus</th>\n",
              "      <th>abund_raw</th>\n",
              "      <th>rel_abund_raw</th>\n",
              "      <th>genome_EC_count</th>\n",
              "      <th>abund_contri</th>\n",
              "      <th>rel_abund_contri</th>\n",
              "      <th>norm_abund_contri</th>\n",
              "      <th>protein_name</th>\n",
              "      <th>...</th>\n",
              "      <th>pathways</th>\n",
              "      <th>hierarchy</th>\n",
              "      <th>metals_involved</th>\n",
              "      <th>metals_consolidated</th>\n",
              "      <th>corrosion_mechanisms</th>\n",
              "      <th>corrosion_relevance_score</th>\n",
              "      <th>corrosion_relevance</th>\n",
              "      <th>has_metal</th>\n",
              "      <th>Category</th>\n",
              "      <th>was_uncharacterized</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>502012</td>\n",
              "      <td>site_27</td>\n",
              "      <td>Afipia</td>\n",
              "      <td>267830.0</td>\n",
              "      <td>28.021013</td>\n",
              "      <td>2</td>\n",
              "      <td>535660.0</td>\n",
              "      <td>56.042026</td>\n",
              "      <td>0.353589</td>\n",
              "      <td>prephenate dehydratase</td>\n",
              "      <td>...</td>\n",
              "      <td>Phenylalanine, tyrosine and tryptophan biosynt...</td>\n",
              "      <td>Amino acid metabolism &gt; Phenylalanine, tyrosin...</td>\n",
              "      <td>potassium; phosphate; iron; sulfur; hydrogen; ...</td>\n",
              "      <td>K; phosphate; H; Mo; Mn; Se; Fe; S</td>\n",
              "      <td>h2_consumption; acid_production; sulfur_metabo...</td>\n",
              "      <td>20.0</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>562124</td>\n",
              "      <td>site_30</td>\n",
              "      <td>Anoxybacillus</td>\n",
              "      <td>32682.0</td>\n",
              "      <td>4.590262</td>\n",
              "      <td>2</td>\n",
              "      <td>65364.0</td>\n",
              "      <td>9.180525</td>\n",
              "      <td>0.054352</td>\n",
              "      <td>corrinoid adenosyltransferase alamin adenosylt...</td>\n",
              "      <td>...</td>\n",
              "      <td>Arginine biosynthesis; Arginine and proline me...</td>\n",
              "      <td>Cardiovascular disease &gt; Viral myocarditis; Ne...</td>\n",
              "      <td>Mo; iron; zinc; hydrogen; selenium; nitrate; n...</td>\n",
              "      <td>H; Zn; Mo; NO3-; Se; Fe; NO2</td>\n",
              "      <td>h2_consumption; direct_eet</td>\n",
              "      <td>14.5</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>472739</td>\n",
              "      <td>site_25</td>\n",
              "      <td>Sphingomonas</td>\n",
              "      <td>1121.0</td>\n",
              "      <td>0.168339</td>\n",
              "      <td>1</td>\n",
              "      <td>1121.0</td>\n",
              "      <td>0.168339</td>\n",
              "      <td>0.001973</td>\n",
              "      <td>phosphoenolpyruvate carboxylase</td>\n",
              "      <td>...</td>\n",
              "      <td>Glycolysis / Gluconeogenesis; Citrate cycle (T...</td>\n",
              "      <td>Energy metabolism &gt; Carbon fixation by Calvin ...</td>\n",
              "      <td>Mo; iron; sulfur; hydrogen; sodium; selenium; ...</td>\n",
              "      <td>H; phosphate; Mo; Se; Fe; S; Na</td>\n",
              "      <td>h2_consumption; acid_production</td>\n",
              "      <td>15.5</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>133617</td>\n",
              "      <td>site_7</td>\n",
              "      <td>Staphylococcus</td>\n",
              "      <td>7736.0</td>\n",
              "      <td>0.878632</td>\n",
              "      <td>1</td>\n",
              "      <td>7736.0</td>\n",
              "      <td>0.878632</td>\n",
              "      <td>0.940890</td>\n",
              "      <td>cystathionine gamma-lyase and homocysteine for...</td>\n",
              "      <td>...</td>\n",
              "      <td>Diterpenoid biosynthesis; Biosynthesis of seco...</td>\n",
              "      <td>Metabolism of terpenoids and polyketides &gt; Dit...</td>\n",
              "      <td>selenium</td>\n",
              "      <td>Se</td>\n",
              "      <td>None</td>\n",
              "      <td>1.5</td>\n",
              "      <td>low</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1136015</td>\n",
              "      <td>site_63</td>\n",
              "      <td>Bradyrhizobium</td>\n",
              "      <td>4161.0</td>\n",
              "      <td>0.613867</td>\n",
              "      <td>8</td>\n",
              "      <td>33288.0</td>\n",
              "      <td>4.910938</td>\n",
              "      <td>0.011502</td>\n",
              "      <td>acetolactate-synthase small subunit</td>\n",
              "      <td>...</td>\n",
              "      <td>Valine, leucine and isoleucine biosynthesis; B...</td>\n",
              "      <td>Metabolism of cofactors and vitamins &gt; Pantoth...</td>\n",
              "      <td>Al3+; Mo; iron; selenium; phosphate</td>\n",
              "      <td>phosphate; Mo; Se; Fe; Al3+</td>\n",
              "      <td>acid_production</td>\n",
              "      <td>10.5</td>\n",
              "      <td>high</td>\n",
              "      <td>True</td>\n",
              "      <td>2</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 24 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e29d1bed-cd29-42bc-a4af-77a37e2e7e32')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-e29d1bed-cd29-42bc-a4af-77a37e2e7e32 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-e29d1bed-cd29-42bc-a4af-77a37e2e7e32');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-a1ac29bd-f763-49dc-86b4-e00af90488a7\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-a1ac29bd-f763-49dc-86b4-e00af90488a7')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-a1ac29bd-f763-49dc-86b4-e00af90488a7 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "sample_contri"
            }
          },
          "metadata": {},
          "execution_count": 127
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#sample_contri  = sample_contri.drop(columns=[\"level_0\", \"index\"], axis=1)\n",
        "# Reset indexes to ensure proper merging\n",
        "#pattern_df = pattern_df.reset_index(drop=True)\n",
        "#sample_contri = sample_contri.drop(columns=\"index\", axis=1)\n",
        "#sample_contri = sample_contri.reset_index(drop=False)"
      ],
      "metadata": {
        "trusted": true,
        "id": "uowLkB5t_FkI"
      },
      "outputs": [],
      "execution_count": 124
    },
    {
      "cell_type": "code",
      "source": [
        "# Reset indexes to ensure proper merging\n",
        "#pattern_df = pattern_df.reset_index(drop=True)\n",
        "#sample_contri = sample_contri.drop(columns=\"index\", axis=1)\n",
        "#sample_contri = sample_contri.reset_index(drop=False)\n",
        "# Perform a single merge to get idx and all other needed columns\n",
        "integrated_results = pd.merge(\n",
        "    pattern_df,\n",
        "    sample_contri[['idx', 'Genus', 'protein_name', 'enzyme_names', 'enzyme_class', 'pathways', 'hierarchy',\n",
        "                 'metals_involved', 'metals_consolidated', 'corrosion_mechanisms',\n",
        "                 'corrosion_relevance_score', 'corrosion_relevance', 'has_metal']],\n",
        "    on=['Genus', 'protein_name'],\n",
        "    how='left'\n",
        ")\n",
        "c\n",
        "integrated_results.head()"
      ],
      "metadata": {
        "trusted": true,
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 193
        },
        "id": "4U2YRQ6K_FkI",
        "outputId": "53488437-b794-4b87-e9bb-36a3f8f1fcc4"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'c' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-132-48a90c2817e9>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'left'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m )\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0mintegrated_results\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'c' is not defined"
          ]
        }
      ],
      "execution_count": 132
    },
    {
      "cell_type": "code",
      "source": [
        "integrated_results = integrated_results.set_index(\"idx\")\n",
        "integrated_results = integrated_results.reset_index()"
      ],
      "metadata": {
        "id": "AsCzkd8CRef6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(f\"h_statistic unique values: {integrated_results['h_statistic'].unique()}, p_value unique values: {integrated_results['p_value'].unique()}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oC3OjCWIQ9QJ",
        "outputId": "01e7d6fd-4695-410d-f1a6-edcf33c5ecad"
      },
      "execution_count": 130,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "h_statistic unique values: [nan], p_value unique values: [nan]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pattern_df.columns"
      ],
      "metadata": {
        "id": "rMufeKiUSlp_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Check how many rows matched during merge\n",
        "before_count = len(pattern_df)\n",
        "after_count = len(integrated_results)\n",
        "print(f\"Rows before merge: {before_count}, after merge: {after_count}\")\n",
        "\n",
        "# Verify idx values are matching\n",
        "print(f\"Unique idx values in pattern_df: {pattern_df['idx'].nunique()}\")\n",
        "print(f\"Unique idx values in ECcontri_Uniprot_enriched: {sample_contri['idx'].nunique()}\")\n",
        "print(f\"Rows with matching idx: {sum(pattern_df['idx'].isin(sample_contri['idx']))}\")\n",
        "# Check if these columns exist in pattern_df before the merge\n",
        "print(\"Columns in pattern_df:\", pattern_df.columns.tolist())\n",
        "print(\"NaN counts in pattern_df:\")\n",
        "print(pattern_df.isna().sum())\n",
        "\n",
        "# Check if these columns exist in eccontri_df before the merge\n",
        "print(\"Columns in ECcontri_Uniprot_enriched:\", sample_contri.columns.tolist())\n",
        "print(\"NaN counts in ECcontri_Uniprot_enriched:\")\n",
        "print(sample_contri.isna().sum())"
      ],
      "metadata": {
        "trusted": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TJruuVHv_FkI",
        "outputId": "443b9e63-7fae-4e91-af64-b14a0a7e7c97"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Rows before merge: 4668, after merge: 5000\n",
            "Unique idx values in pattern_df: 4668\n",
            "Unique idx values in ECcontri_Uniprot_enriched: 5000\n",
            "Rows with matching idx: 4668\n",
            "Columns in pattern_df: ['idx', 'Genus', 'protein_name', 'h_statistic', 'p_value', 'pattern', 'presence_pattern', 'category_count', 'categories', 'corr', 'mean_cat1', 'median_cat1', 'std_cat1', 'count_cat1', 'sum_cat1', 'mean_cat2', 'median_cat2', 'std_cat2', 'count_cat2', 'sum_cat2', 'mean_cat3', 'median_cat3', 'std_cat3', 'count_cat3', 'sum_cat3', 'effect_size_1_to_2', 'effect_size_2_to_3', 'total_abundance', 'log2fc_cat2_vs_cat1', 'log2fc_cat3_vs_cat1', 'log2fc_cat3_vs_cat2', 'idx_y']\n",
            "NaN counts in pattern_df:\n",
            "idx                       0\n",
            "Genus                     0\n",
            "protein_name              0\n",
            "h_statistic            4668\n",
            "p_value                4668\n",
            "pattern                   0\n",
            "presence_pattern          0\n",
            "category_count            0\n",
            "categories                0\n",
            "corr                      0\n",
            "mean_cat1              3370\n",
            "median_cat1            3370\n",
            "std_cat1               4668\n",
            "count_cat1             3370\n",
            "sum_cat1               3370\n",
            "mean_cat2              2576\n",
            "median_cat2            2576\n",
            "std_cat2               4668\n",
            "count_cat2             2576\n",
            "sum_cat2               2576\n",
            "mean_cat3              3183\n",
            "median_cat3            3183\n",
            "std_cat3               4668\n",
            "count_cat3             3183\n",
            "sum_cat3               3183\n",
            "effect_size_1_to_2     4594\n",
            "effect_size_2_to_3     4577\n",
            "total_abundance           0\n",
            "log2fc_cat2_vs_cat1    4594\n",
            "log2fc_cat3_vs_cat1    4616\n",
            "log2fc_cat3_vs_cat2    4577\n",
            "idx_y                     0\n",
            "dtype: int64\n",
            "Columns in ECcontri_Uniprot_enriched: ['idx', 'Sites', 'Genus', 'abund_raw', 'rel_abund_raw', 'genome_EC_count', 'abund_contri', 'rel_abund_contri', 'norm_abund_contri', 'protein_name', 'uniprot_id', 'EC', 'enzyme_names', 'enzyme_class', 'pathways', 'hierarchy', 'metals_involved', 'metals_consolidated', 'corrosion_mechanisms', 'corrosion_relevance_score', 'corrosion_relevance', 'has_metal', 'Category', 'was_uncharacterized']\n",
            "NaN counts in ECcontri_Uniprot_enriched:\n",
            "idx                            0\n",
            "Sites                          0\n",
            "Genus                          0\n",
            "abund_raw                      0\n",
            "rel_abund_raw                  0\n",
            "genome_EC_count                0\n",
            "abund_contri                   0\n",
            "rel_abund_contri               0\n",
            "norm_abund_contri              0\n",
            "protein_name                   0\n",
            "uniprot_id                   918\n",
            "EC                             0\n",
            "enzyme_names                   0\n",
            "enzyme_class                  16\n",
            "pathways                     229\n",
            "hierarchy                    305\n",
            "metals_involved                0\n",
            "metals_consolidated            0\n",
            "corrosion_mechanisms         502\n",
            "corrosion_relevance_score      0\n",
            "corrosion_relevance            0\n",
            "has_metal                      0\n",
            "Category                       0\n",
            "was_uncharacterized            0\n",
            "dtype: int64\n"
          ]
        }
      ],
      "execution_count": 131
    }
  ]
}